{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for loading in Parse Bio AnnData object and spliting into two separate objects for each study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # These packages are pre-installed on Google Colab, but are included here to facilitate running this notebook locally\n",
    "# !pip install --quiet matplotlib\n",
    "# !pip install --quiet scikit-learn\n",
    "# !pip install --quiet numpy\n",
    "# !pip install --quiet scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "docker 6.1.3 requires websocket-client>=0.32.0, which is not installed.\u001b[0m\u001b[31m\n",
      "\u001b[0mCPU times: user 599 ms, sys: 80.8 ms, total: 679 ms\n",
      "Wall time: 30.8 s\n"
     ]
    }
   ],
   "source": [
    "# %%time\n",
    "# # Install scanpy and other packages needed for single-cell RNA-seq analysis\n",
    "# !pip install --quiet scanpy python-igraph louvain pybiomart #MulticoreTSNE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import anndata\n",
    "import matplotlib\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from scipy import sparse, io\n",
    "from tqdm import tqdm\n",
    "\n",
    "matplotlib.rcParams.update({'font.size': 12})\n",
    "%config InlineBackend.figure_format = 'retina'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading AnnData Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = \"/resnick/groups/mthomson/jboktor/split-pipe_workflow/combined_parallel/all-sample/DGE_filtered\"\n",
    "adata_path = os.path.join(data_path, 'anndata.h5ad')\n",
    "analysis_path = \"/central/groups/MazmanianLab/joeB/snRNAseq_analysis/WILDRxSPF_brains\"\n",
    "output_path = os.path.join(analysis_path, \"data/input/parse_bio_anndata\")\n",
    "os.makedirs(output_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 719168 × 78298\n",
       "    obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count'\n",
       "    var: 'gene_id', 'gene_name', 'genome'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = anndata.read_h5ad(adata_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gnai3', 'Pbsn', 'Cdc45', 'H19', 'Scml2', 'Apoh', 'Narf', 'Cav2',\n",
       "       'Klf6', 'Scmh1',\n",
       "       ...\n",
       "       'Gm55949', 'Gm56260', 'Gm56355', 'Gm54430', 'Gm54490', 'Gm56371',\n",
       "       'Gm23510', 'Gm22711', 'Gm55627', 'Gm54807'],\n",
       "      dtype='object', length=78298)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var_names.tolist()\n",
    "# adata.var_names = [name.replace('_GRCm39', '') for name in adata.var_names]\n",
    "adata.var_names = [name.split('_GRCm39')[0] for name in adata.var_names]\n",
    "adata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene IDs:\n",
      "Total count: 78298\n",
      "Unique count: 78298\n",
      "Number of duplicates: 0\n",
      "\n",
      "Gene Names:\n",
      "Total count: 78298\n",
      "Unique count: 77981\n",
      "Number of duplicates: 317\n",
      "\n",
      "Duplicate gene_names:\n",
      "                  gene_id      gene_name  genome\n",
      "54335  ENSMUSG00000121691  1600029O15Rik  GRCm39\n",
      "16406  ENSMUSG00000057818  1600029O15Rik  GRCm39\n",
      "56229  ENSMUSG00000123653  1700016A09Rik  GRCm39\n",
      "37421  ENSMUSG00000102737  1700016A09Rik  GRCm39\n",
      "54513  ENSMUSG00000121875  1700020L13Rik  GRCm39\n",
      "...                   ...            ...     ...\n",
      "54151  ENSMUSG00000121502         Zfp939  GRCm39\n",
      "42083  ENSMUSG00000108043           Zim3  GRCm39\n",
      "54516  ENSMUSG00000121878           Zim3  GRCm39\n",
      "47252  ENSMUSG00000113593         Zp4-ps  GRCm39\n",
      "54026  ENSMUSG00000121372         Zp4-ps  GRCm39\n",
      "\n",
      "[629 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# Load the gene mapping file\n",
    "gene_mapping = pd.read_csv(\"/resnick/groups/mthomson/jboktor/split-pipe_workflow/combined_parallel/all-sample/DGE_filtered/all_genes.csv\")\n",
    "gene_mapping\n",
    "\n",
    "# Check for duplicates in gene_id\n",
    "n_gene_ids = len(gene_mapping['gene_id'])\n",
    "n_unique_gene_ids = len(gene_mapping['gene_id'].unique())\n",
    "print(f\"\\nGene IDs:\")\n",
    "print(f\"Total count: {n_gene_ids}\")\n",
    "print(f\"Unique count: {n_unique_gene_ids}\")\n",
    "print(f\"Number of duplicates: {n_gene_ids - n_unique_gene_ids}\")\n",
    "\n",
    "if n_gene_ids != n_unique_gene_ids:\n",
    "    print(\"\\nDuplicate gene_ids:\")\n",
    "    print(gene_mapping[gene_mapping['gene_id'].duplicated(keep=False)].sort_values('gene_id'))\n",
    "\n",
    "# Check for duplicates in gene_name\n",
    "n_gene_names = len(gene_mapping['gene_name'])\n",
    "n_unique_gene_names = len(gene_mapping['gene_name'].unique())\n",
    "print(f\"\\nGene Names:\")\n",
    "print(f\"Total count: {n_gene_names}\")\n",
    "print(f\"Unique count: {n_unique_gene_names}\")\n",
    "print(f\"Number of duplicates: {n_gene_names - n_unique_gene_names}\")\n",
    "\n",
    "if n_gene_names != n_unique_gene_names:\n",
    "    print(\"\\nDuplicate gene_names:\")\n",
    "    print(gene_mapping[gene_mapping['gene_name'].duplicated(keep=False)].sort_values('gene_name'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All Ensemble ID's are unique but some gene_names are duplicates - must be mapped to different ensemble IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Swapping in gene symbol names for OG Ensembl IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of names: 78298\n",
      "Number of unique names: 78298\n",
      "Are all names unique? True\n"
     ]
    }
   ],
   "source": [
    "adata.var_names = gene_mapping['gene_id']\n",
    "\n",
    "n_total = len(adata.var_names)\n",
    "n_unique = len(set(adata.var_names))\n",
    "\n",
    "print(f\"Total number of names: {n_total}\")\n",
    "print(f\"Number of unique names: {n_unique}\")\n",
    "print(f\"Are all names unique? {n_total == n_unique}\")\n",
    "\n",
    "# If there are duplicates, we can find them:\n",
    "if n_total != n_unique:\n",
    "    duplicates = pd.Series(adata.var_names)[pd.Series(adata.var_names).duplicated()].unique()\n",
    "    print(\"\\nDuplicate names found:\")\n",
    "    print(duplicates)\n",
    "    print(f\"\\nNumber of duplicate names: {len(duplicates)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Saving AnnData with unique cellxgene matrix with unique ensemble IDs for Cell Type Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/central/groups/MazmanianLab/joeB/snRNAseq_analysis/WILDRxSPF_brains/data/input/parse_bio_anndata/anndata_ensembl.h5ad\n"
     ]
    }
   ],
   "source": [
    "print(os.path.join(output_path, 'anndata_ensembl.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(os.path.join(output_path, 'anndata_ensembl.h5ad'))\n",
    "# Saving the AnnData object with Ensembl IDs for Cell Type Mapping\n",
    "adata.write_h5ad(os.path.join(output_path, 'anndata_ensembl.h5ad'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filtering samples for each study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean load of data\n",
    "adata = anndata.read_h5ad(adata_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Filter cells\n",
    "sc.pp.filter_cells(adata, min_counts=1000, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3, min_counts=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "View of AnnData object with n_obs × n_vars = 301709 × 78298\n",
      "    obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count'\n",
      "    var: 'gene_id', 'gene_name', 'genome'\n",
      "ASO11\n",
      "ASO13\n",
      "ASO15\n",
      "ASO16\n",
      "ASO17\n",
      "ASO20\n",
      "WT11\n",
      "WT12\n",
      "WT13\n",
      "WT14\n",
      "WT15\n",
      "WT17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/central/groups/MazmanianLab/joeB/software/mambaforge/envs/anndata/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:168: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs[\"n_counts\"] = number\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 300016 × 51948\n",
      "    obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'n_counts', 'n_genes'\n",
      "    var: 'gene_id', 'gene_name', 'genome', 'n_cells', 'n_counts'\n"
     ]
    }
   ],
   "source": [
    "manxuan_adata  = adata[adata.obs['sample'].str.contains('ASO|WT', regex=True)]\n",
    "print(manxuan_adata)\n",
    "for sample in manxuan_adata.obs['sample'].unique():\n",
    "    print(sample)\n",
    "\n",
    "sc.pp.filter_cells(manxuan_adata, min_counts=1000)  # keep cells with ≥1,000 total counts :contentReference[oaicite:0]{index=0}\n",
    "sc.pp.filter_cells(manxuan_adata, min_genes=200)    # keep cells with ≥200 genes detected :contentReference[oaicite:1]{index=1}\n",
    "sc.pp.filter_genes(manxuan_adata, min_cells=20)      # keep genes seen in ≥20 cells :contentReference[oaicite:2]{index=2}\n",
    "sc.pp.filter_genes(manxuan_adata, min_counts=10)    # keep genes with ≥10 total counts :contentReference[oaicite:3]{index=3}\n",
    "\n",
    "print(manxuan_adata)\n",
    "manxuan_adata.write_h5ad(os.path.join(output_path, 'aso_wt_data.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "View of AnnData object with n_obs × n_vars = 417459 × 78298\n",
      "    obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count'\n",
      "    var: 'gene_id', 'gene_name', 'genome'\n",
      "SPF_10_HYP\n",
      "SPF_11_AMY\n",
      "SPF_1D4_HYP\n",
      "SPF_3D4_AMY\n",
      "SPF_3D4_HYP\n",
      "SPF_6_AMY\n",
      "SPF_6_HYP\n",
      "SPF_7_AMY\n",
      "SPF_8_AMY\n",
      "SPF_8_HYP\n",
      "SPF_9_AMY\n",
      "SPF_9_HYP\n",
      "WildR_10_AMY\n",
      "WildR_10_HYP\n",
      "WildR_11_AMY\n",
      "WildR_11_HYP\n",
      "WildR_3D4_HYP\n",
      "WildR_5_AMY\n",
      "WildR_6_AMY\n",
      "WildR_6_HYP\n",
      "WildR_7_AMY\n",
      "WildR_7_HYP\n",
      "WildR_9_AMY\n",
      "WildR_9_HYP\n",
      "AnnData object with n_obs × n_vars = 414591 × 56721\n",
      "    obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'n_counts', 'n_genes'\n",
      "    var: 'gene_id', 'gene_name', 'genome', 'n_cells', 'n_counts'\n"
     ]
    }
   ],
   "source": [
    "jess_adata = adata[~adata.obs['sample'].str.contains('ASO|WT', regex=True)]\n",
    "print(jess_adata)\n",
    "for sample in jess_adata.obs['sample'].unique():\n",
    "    print(sample)\n",
    "    \n",
    "sc.pp.filter_cells(jess_adata, min_counts=1000)  # keep cells with ≥1,000 total counts\n",
    "sc.pp.filter_cells(jess_adata, min_genes=200)    # keep cells with ≥200 genes detected\n",
    "sc.pp.filter_genes(jess_adata, min_cells=20)      # keep genes seen in ≥20 cells\n",
    "sc.pp.filter_genes(jess_adata, min_counts=10)    # keep genes with ≥10 total counts\n",
    "\n",
    "print(jess_adata)\n",
    "jess_adata.write_h5ad(os.path.join(output_path, 'wildrxspf_data.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anndata",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
