---
title: "WildR Brain Tissue Spatial Transcriptomics Analysis"
editor: source
author: "Joe Boktor"
date: '2024-04-13'
format: 
  html:
    font-family: helvetica neue
    page-layout: full
    toc: true
    toc-location: left
    toc-depth: 5
    self-contained: true
    code-fold: false
    code-tools: true
    fig-align: center
    grid:
      sidebar-width: 200px
      body-width: 1100px
      margin-width: 200px
      gutter-width: 1.5em
---

# Background

Here we analyze

# Analysis Setup

Environment setup.

```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(spdep)
library(Voyager)
library(SingleCellExperiment)
library(SpatialExperiment)
library(SpatialFeatureExperiment)
library(batchelor)
library(scran)
library(scater)
library(bluster)
library(fossil)
library(glue)
library(Seurat)
library(scCustomize)
library(glmGamPoi)
library(spatialLIBD)
library(reticulate)
library(SeuratDisk)
library(rhdf5)
library(anndata)
library(grid)
library(biomaRt)
library(SingleR)
library(clusterProfiler)
library(org.Mm.eg.db)
library(batchtools)

# stats 
library(lme4)
library(broom)

# parallelization
library(BiocParallel)
library(future)
library(furrr)

# plotting
library(RColorBrewer)
library(ggsci)
library(Matrix)
library(plotly)
library(DT)
library(gtsummary)
library(aplot)
library(pheatmap)
library(patchwork)
library(VennDiagram)
library(viridis)

# setting paths
homedir <- "/central/groups/mthomson/jboktor"
wkdir <- glue("{homedir}/spatial_genomics/jess_2024-01-23")
source(glue("{wkdir}/notebooks/R_scripts/_misc_functions.R"))

# 1.5 GB limit 1500*1024^2 = 1572864000
# 12 GB limit = 12582912000 * 8  
options(future.globals.maxSize= 12582912000)
future::plan("multisession", workers = 24)

```

### Loading and formatting seqFISH data into Seurat objects

```{r, eval = FALSE}

#' Function to load and process cell-by-gene matrix and metadata
load_ROI_seurat_trim <- function(cbg_path, meta_path) {
  require(Matrix)
  require(Seurat)
  require(magrittr)

  cell_by_gene_matrix <-
    read.csv(cbg_path, row.names = 1) %>%
    dplyr::rename_all(~ toupper(.)) %>%
    t()

  # select for cells with at least 40 counts
  quality_cells_min <- colSums(cell_by_gene_matrix) >= 40
  quality_cells_max <- colSums(cell_by_gene_matrix) <= 3000
  quality_cells <- quality_cells_min & quality_cells_max
  perc_remain <- sum(quality_cells) / length(quality_cells) * 100
  message(
    "\n\nCells below 40 total counts: ", sum(!quality_cells_min),
    "\nCells above 3000 total counts: ", sum(!quality_cells_max),
    "\nCells remaining: ", sum(quality_cells),
    "\nRetaining ", round(perc_remain, 2), "% of cells\n"
  )
  cell_by_gene_matrix <- cell_by_gene_matrix[, quality_cells]
  cell_by_gene_matrix %<>% as("dgCMatrix")

  # load in cell-coordinates / metadata
  cell_attributes <- read.csv(meta_path)
  cell_attributes <- cell_attributes[quality_cells, ]

  # Check the dimensions and structure of your input data
  print(dim(cell_by_gene_matrix))
  print(dim(cell_attributes))
  
  # Create Seurat object
  seur <-
    Seurat::CreateSeuratObject(
      counts = cell_by_gene_matrix,
      project = "SeqFISH"
    ) %>%
    Seurat::AddMetaData(metadata = cell_attributes)

  # add assay and make active
  seur@assays[["Spatial"]] <-
    CreateAssayObject(counts = cell_by_gene_matrix)
  seur@active.assay <- "Spatial"
  DefaultAssay(seur) <- "Spatial"

  seur <- UpdateSeuratObject(seur)

  return(seur)
}

roi_list <- paste0("roi", 1:4)
run_list <- paste0("run", 1:4)
data_dir <- glue("{wkdir}/data/input/SpatialAnalysis_2024-07")
seurat_list <- list()
group_list <- list(
  roi1_run1 = "spf",
  roi2_run1 = "wildr",
  roi3_run1 = "wildr",
  roi4_run1 = "spf",
  roi1_run2 = "wildr",
  roi2_run2 = "spf",
  roi3_run2 = "spf",
  roi4_run2 = "wildr",
  roi1_run3 = "spf",
  roi2_run3 = "wildr",
  roi3_run3 = "wildr",
  roi4_run3 = "spf",
  roi1_run4 = "spf",
  roi2_run4 = "wildr",
  roi3_run4 = "wildr",
  roi4_run4 = "spf"
)
bregma_list <- list(
  roi1_run1 = -1.85,
  roi2_run1 = -1.95,
  roi3_run1 = -2.45,
  roi4_run1 = -1.85,
  roi1_run2 = -1.65,
  roi2_run2 = -1.65,
  roi3_run2 = -2.4,
  roi4_run2 = -1.85,
  roi1_run3 = -1.85,
  roi2_run3 = -2.15,
  roi3_run3 = -1.95,
  roi4_run3 = -1.95,
  roi1_run4 = -1.35,
  roi2_run4 = -1.85,
  roi3_run4 = -1.85,
  roi4_run4 = -1.75
)
block_list <- list(
  run1 = "block2",
  run2 = "block3",
  run3 = "block1",
  run4 = "block4"
)

for (roi in roi_list) {
  for (run in run_list) {
    sample <- glue("{roi}_{run}")
    block_var <- block_list[[run]]
    sample_dir <- glue("export_{run}_{roi}_{block_var}_july2024")
    print(glue("Loading: {sample}"))

    cbg_path <- glue(
      "{data_dir}/{sample_dir}/{sample_dir}/",
      "cell_by_gene_matrix_{roi}_{block_var}.csv"
    )
    meta_path <- glue(
      "{data_dir}/{sample_dir}/{sample_dir}/",
      "cell_attributes_{roi}_{block_var}.csv"
    )

    if (!file.exists(cbg_path)) {
      print(glue("ERROR \n {cbg_path} \ndoes not exist!"))
      next
    }

    seurat_list[[sample]] <- load_ROI_seurat_trim(
      cbg_path = cbg_path,
      meta_path = meta_path
    )

    # adding metadata
    seurat_list[[sample]]$roi <- roi
    seurat_list[[sample]]$run <- run
    seurat_list[[sample]]$block <- block_var
    seurat_list[[sample]]$group <- group_list[[sample]]
    seurat_list[[sample]]$bregma <- bregma_list[[sample]]
  
  }
}

merged_rois <- Merge_Seurat_List(
  seurat_list,
  add.cell.ids = names(seurat_list),
  merge.data = TRUE,
  project = "SeuratProject"
)

saveRDS(
    merged_rois,
    glue("{wkdir}/data/interim/merged_roi_seurat_prenorm_{Sys.Date()}.rds")
)

```


```{r, eval = FALSE}

merged_rois <- readRDS(
    glue("{wkdir}/data/interim/merged_roi_seurat_prenorm_2024-07-15.rds")
)

merged_rois %<>%
  UpdateSeuratObject() %>%
  SCTransform(assay = "Spatial")

# Run umap and generate clusters
merged_rois %<>%
  RunPCA() %>%
  FindNeighbors(dims = 1:30) %>%
  FindClusters() %>% 
  RunUMAP(dims = 1:20)

merged_rois@meta.data %>% glimpse

saveRDS(
    merged_rois,
    glue("{wkdir}/data/interim/merged_roi_seurat_{Sys.Date()}.rds")
)

meta_df <- merged_rois@meta.data

saveRDS(
  meta_df,
  glue("{wkdir}/data/interim/merged_roi_seurat_metadata_{Sys.Date()}.rds")
)

```

Functions to identify boundaries of problematic regions

```{r, eval = FALSE}
#' Function to annotate slice artifacts using point coordinates
#' example inputs
#' 
#' points_1 = list(
#'  "boundary1" = c(2000, 500), 
#'  "boundary2" = c(2500, 3800)
#'  )
#' points_2 = list(
#'  "boundary1" = c(3000, 0), 
#'  "boundary2" = c(3000, 3500)
#'  )
#' slice_ids = c("WILDR run3 roi2", "WILDR run3 roi2")
annotate_slice_artifacts <- function(seurat_obj, slice_ids, points_1, points_2) {
  require(magrittr)
  require(dplyr)
  
  if (length(points_1) != length(points_2)) {
  stop("Length of points_1 and points_2 must be equal")
  }
  
  meta <- seurat_obj@meta.data %>%
    mutate(artifact_region = "valid")
  
  for (bound_n in 1:length(slice_ids)){
    if (points_1[[bound_n]][1] > points_2[[bound_n]][1]){
      stop("Boundary 1 x-coordinate must be less than Boundary 2")
    }
    else if (points_1[[bound_n]][2] > points_2[[bound_n]][2]){
      stop("Boundary 1 y-coordinate must be smaller than Boundary 2")
    }
    # message(glue("Boundary {bound_n} is valid")

    meta %<>%
      mutate(artifact_region = case_when(
        artifact_region == "artifact" ~ "artifact",
        slice_id == slice_ids[bound_n] &
        (center_x > points_1[[bound_n]][1] & center_x < points_2[[bound_n]][1]) &
        (center_y < points_2[[bound_n]][2] & center_y > points_1[[bound_n]][2]) ~ "artifact",
        TRUE ~ "valid"
      ))
  }
  meta$artifact_region %>% table() %>% print()
  return(meta)
}

plot_slice_artifacts <- function(metadata_df) {
    p <- metadata_df %>%
    arrange(artifact_region) %>%
    ggplot(aes(x = center_x, y = center_y)) +
    geom_point(aes(color = artifact_region), size = 0.2, alpha = 0.2) +
    facet_wrap(~slice_id, ncol = 4) +
    labs(color = "Artifact Region") +
    coord_fixed() +
    scale_y_reverse() +
    scale_color_manual(values = c("valid" = "grey", "artifact" = "red")) +
    theme_bw()
  return(p)
}

merged_rois$slice_id %>% unique

# plot_slice_artifacts 
meta_df <- annotate_slice_artifacts(
  seurat_obj = merged_rois,
  slice_ids = c(
    "SPF run1 roi1",
    "WILDR run3 roi2", 
    "WILDR run3 roi2",
    "SPF run3 roi4"
    ),
  points_1 = list(
    "SPF r1r1" = c(0, 0),
    "WR r3r2 topband" = c(0, 0), 
    "WR r3r2 bottomband" = c(0, 40000),
    "SPF r3r4 topband" = c(0, 5500)
  ),
  points_2 = list(
    "SPF r1r1" = c(50000, 6000),
    "WR r3r2 topband" = c(50000, 9000), 
    "WR r3r2 bottomband" = c(40000, 42500),
    "SPF r3r4 topband" = c(50000, 8000)
    )
)

p <- plot_slice_artifacts(meta_df)
ggsave(
  filename = glue("{wkdir}/figures/slice_artifacts_{Sys.Date()}.png"),
  p,
  width = 16, height = 16
)

```

Quality Control Summary Stat Tables
```{r, eval = TRUE}
meta_df <- readRDS(
  glue("{wkdir}/data/interim/merged_roi_seurat_metadata_2024-07-15.rds")
)
meta_df %<>% 
  mutate(slice_id = glue("{group} {roi} {run}"))

meta_df %>%
  group_by(slice_id, group, bregma) %>%
  summarise(n_cells = n()) %>%
  DT::datatable()

meta_df %>%
  tbl_summary(include = c(slice_id)) %>%
  modify_header(label = "**Cells**") %>%
  bold_labels()

meta_df %>%
  tbl_summary(include = c(group), by = "run") %>%
  modify_header(label = "**Cells**") %>%
  bold_labels() %>% 
  add_overall()

meta_df %>%
  tbl_summary(include = c(nCount_RNA, nFeature_RNA), by = "group") %>%
  modify_header(label = "**Cells**") %>%
  bold_labels() %>%
  add_overall() %>%
  add_p()

```



Visualizing distributions of nCount and nFeature

```{r, eval = FALSE}
p_hist_nCount_group <- meta_df %>%
  ggplot(aes(nCount_RNA, fill = group)) +
  geom_histogram(bins = 100, alpha=0.5, position="identity") +
  scale_fill_nejm() +
  labs(x = "Transcript Counts per Cell", fill = "Group") +
  theme_bw()

p_hist_nCount_run <- meta_df %>%
  ggplot(aes(nCount_RNA, fill = run)) +
  geom_histogram(bins = 100, alpha = 0.5, position = "identity") +
  scale_fill_npg() +
  labs(x = "Transcript Counts per Cell", fill = "Run") +
  theme_bw()

p_hist_nfeat_group <- meta_df %>%
  ggplot(aes(nFeature_RNA, fill = group)) +
  geom_histogram(bins = 50, alpha = 0.5, position = "identity") +
  scale_fill_nejm() +
    labs(x = "Unique Genes per Cell", fill = "Group") +
  theme_bw()

p_hist_nfeat_run <- meta_df %>%
  ggplot(aes(nFeature_RNA, fill = run)) +
  geom_histogram(bins = 50, alpha = 0.5, position = "identity") +
  scale_fill_npg() +
  labs(x = "Unique Genes per Cell", fill = "Run") +
  theme_bw()


p_ncount <- insert_bottom(p_hist_nCount_group, p_hist_nCount_run)
p_nfeat <- insert_bottom(p_hist_nfeat_group, p_hist_nfeat_run)

ggsave(
  glue("{wkdir}/figures/merged_roi_nCount_hist_{Sys.Date()}.png"),
  p_ncount,
  width = 5, height = 6
)

ggsave(
  glue("{wkdir}/figures/merged_roi_nFeat_hist_{Sys.Date()}.png"),
  p_nfeat,
  width = 5, height = 6
)

```

![Transcript count distributions by group and run](../figures/merged_roi_nCount_hist_2024-06-23.png) 


![Unique gene count distributions by group and run](../figures/merged_roi_nFeat_hist_2024-06-23.png)


Examining count distribution of genes (serial and barcoded)

```{r, eval = FALSE}

merged_rois <- readRDS(
    glue("{wkdir}/data/interim/merged_roi_seurat_prenorm_2024-07-14.rds")
)

neuromapper_serial_genes <- read.csv(
  glue("{wkdir}/data/input/NeuroMapper_Channel2_gene_list.csv"),
  row.names = NULL
)

serial_genes <- toupper(neuromapper_serial_genes$gene_name)
counts_df <- GetAssayData(merged_rois, layer = 'counts')
gene_counts <- rowSums(counts_df)
gene_hyb_type <- data.frame(
  gene = names(gene_counts),
  count = unname(gene_counts)
) %>%
  mutate(
    gene_type = if_else(gene %in% serial_genes, "serial", "barcoded")
  ) %>%
  glimpse()

p_hybtype_dist <- gene_hyb_type %>%
  ggplot(aes(count)) +
  geom_histogram(bins = 20, aes(fill = gene_type)) +
  scale_x_log10() +
  theme_bw()

ggsave(
  glue("{wkdir}/figures/gene_hyb-type-distribution_{Sys.Date()}.png"),
  p_hybtype_dist,
  width = 4, height = 3
)


```


![Gene count distribution by hybridization type](../figures/gene_hyb-type-distribution_2024-06-23.png)



### Exporting data for osNMF analysis

```{r, eval = FALSE}
seurat_data <- merged_rois %>% 
  #subset(slice_id == "roi1_run1") %>%
  GetAssayData(layer = "data")
filtered_data <- as.matrix(seurat_data)  # Convert to matrix

write.csv(
  filtered_data,
  glue("{wkdir}/data/interim/seurat_data_{Sys.Date()}.csv")
)
```


```{r, eval = TRUE}
sessionInfo()
```