---
title: "WildR Brain Tissue Spatial Transcriptomics Analysis"
editor: source
author: "Joe Boktor"
date: '2024-04-13'
format: 
  html:
    font-family: helvetica neue
    page-layout: full
    toc: true
    toc-location: left
    toc-depth: 5
    self-contained: true
    code-fold: false
    code-tools: true
    fig-align: center
    grid:
      sidebar-width: 200px
      body-width: 1100px
      margin-width: 200px
      gutter-width: 1.5em
---

# Background

Here we analyze

# Analysis Setup

Environment setup.

```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(spdep)
library(Voyager)
library(SingleCellExperiment)
library(SpatialExperiment)
library(SpatialFeatureExperiment)
library(batchelor)
library(scran)
library(scater)
library(bluster)
library(fossil)
library(glue)
library(Seurat)
library(scCustomize)
library(glmGamPoi)
library(spatialLIBD)
library(reticulate)
library(SeuratDisk)
library(rhdf5)
library(anndata)
library(grid)
library(biomaRt)
library(SingleR)
library(clusterProfiler)
library(org.Mm.eg.db)
library(batchtools)

# stats 
library(lme4)
library(broom)

# parallelization
library(BiocParallel)
library(future)
library(furrr)

# plotting
library(RColorBrewer)
library(ggsci)
library(Matrix)
library(plotly)
library(DT)
library(gtsummary)
library(aplot)
library(pheatmap)
library(patchwork)
library(VennDiagram)
library(viridis)

# setting paths
homedir <- "/central/groups/mthomson/jboktor"
wkdir <- glue("{homedir}/spatial_genomics/jess_2024-01-23")
source(glue("{wkdir}/notebooks/R_scripts/_misc_functions.R"))
use_condaenv(condaenv = 'spatialomics', required = TRUE)

# 12gb  limit (1500*1024^2 = 1572864000) * 8
options(future.globals.maxSize= 12582912000)
future::plan("multisession", workers = 24)

```


# osNMF Analysis of raw count data

Prepping R objects for python cells

```{r, eval=FALSE}
merged_rois <- readRDS(
    glue("{wkdir}/data/interim/merged_roi_seurat_2024-07-15.rds")
)
seurat_data <- merged_rois %>% 
  GetAssayData(layer = "data")
filtered_data <- as.matrix(seurat_data)  # Convert to matrix

```

Running staNMF on all cell types - using a prespecified k (11)

```{python}
import numpy as np
from sklearn.decomposition import NMF as sklearn_nmf
import sys
sys.path.append('/central/groups/mthomson/jboktor/spatial_genomics/osNMF/osNMF')
from knn_impute import kNN_imputation
from morans_index import MoranIforPPs
from other_functions import show_graph_with_labels, filter_genes, weighted_correlation
from sklearn_nmf import sklearn_nmf
from stability import findcorrelation, HungarianError, amariMaxError, instability

count_mat = r.filtered_data
nmf = sklearn_nmf(n_components=11, l1_ratio=1, alpha=0, random_state=1, max_iter=1000)
nmf.fit(np.maximum(count_mat, 0))

PPs = nmf.components_
coeffs = nmf.transform(np.maximum(count_mat, 0))

# convert into R objects
r.PPs = PPs
r.coeffs = coeffs
```

Convert the results back into R objects

```{r}
PPs_r <- matrix(PPs, nrow = dim(PPs)[1], ncol = dim(PPs)[2]) %>%
  t() %>%
  set_colnames(paste0("osNMF_", 1:dim(PPs)[1]))
coeffs_r <- matrix(coeffs, nrow = dim(coeffs)[1], ncol = dim(coeffs)[2])

cat("PPs shape:", dim(PPs_r), "\n")
cat("Coeffs shape:", dim(coeffs_r), "\n")

saveRDS(
  PPs_r,
  glue("{wkdir}/data/interim/osNMF/osNMF_PPs_K11_{Sys.Date()}.rds")
)
saveRDS(
  coeffs_r,
  glue("{wkdir}/data/interim/osNMF/osNMF_coeffs_K11_{Sys.Date()}_.rds")
)

```

## Plotting seqFISH NMF DIMS

```{r, eval=FALSE}
# adding NMF results to Seurat object
merged_rois <- AddMetaData(
  object = merged_rois,
  metadata = PPs_r
)
merged_rois$slice_id <- 
    glue("{merged_rois$run}_{merged_rois$roi}")
merged_rois@meta.data %>% glimpse

nmdf_dims <- merged_rois@meta.data %>% 
  colnames() %>% 
  keep(grepl("osNMF", .))

for (d in paste0("osNMF_", 1:length(nmdf_dims))){
  message(glue("Plotting: {d}"))
  p <- merged_rois@meta.data %>%
    rownames_to_column(var = "cell_id") %>%
    arrange(!!sym(d)) %>%
    mutate(cell_id, factor(cell_id)) %>%
    arrange(group) %>%
    mutate(slice_id = paste0(toupper(group), " ", slice_id)) %>%
    mutate(slice_id, factor(slice_id)) %>%
    ggplot(aes(center_x, -center_y, color = !!sym(d))) +
    geom_point(size = 0.3) +
    scale_color_viridis() +
    coord_fixed() +
    facet_wrap(~slice_id, ncol = 4) +
    theme_bw()

  ggsave(
    glue("{wkdir}/figures/osNMF_K11/{Sys.Date()}/K11_{d}_{Sys.Date()}.png"),
    p,
    width = 16, height = 16
  )
}

```


# Running staNMF on ABCA reference slices

```{r}
ref_seur_trim <- readRDS(
  glue("{wkdir}/data/interim/ABCA_1&2_z-6-to-8.5_seurat.rds")
)
ref_seurat_counts <- GetAssayData(ref_seur_trim, 
  assay = "RNA", 
  slot = "counts"
)
ref_counts <- as.matrix(seurat_data)

```


```{python}
import numpy as np
from sklearn.decomposition import NMF as sklearn_nmf
import sys
sys.path.append('/central/groups/mthomson/jboktor/spatial_genomics/osNMF/osNMF')
from knn_impute import kNN_imputation
from morans_index import MoranIforPPs
from other_functions import show_graph_with_labels, filter_genes, weighted_correlation
from sklearn_nmf import sklearn_nmf
from stability import findcorrelation, HungarianError, amariMaxError, instability

ref_count_mat = r.ref_counts.toarray()

nmf = sklearn_nmf(n_components=11, l1_ratio=1, alpha=0, random_state=1, max_iter=5000)
nmf.fit(np.maximum(ref_count_mat, 0))

PPs = nmf.components_
coeffs = nmf.transform(np.maximum(ref_count_mat, 0))

# convert into R objects
r.PPs_ref = PPs
r.coeffs_ref = coeffs
```


```{r}
# Convert the results back into R objects
ref_PPs_r <- matrix(PPs_ref, nrow = dim(PPs_ref)[1], ncol = dim(PPs_ref)[2]) %>%
  t() %>%
  set_colnames(paste0("osNMF_", 1:dim(PPs_ref)[1]))

ref_coeffs_r <- matrix(coeffs_ref, nrow = dim(coeffs_ref)[1], ncol = dim(coeffs_ref)[2])

cat("PPs shape:", dim(ref_PPs_r), "\n")
cat("Coeffs shape:", dim(ref_coeffs_r), "\n")

saveRDS(
  ref_PPs_r,
  glue("{wkdir}/data/interim/osNMF/ABCA1&2_osNMF_PPs_K11_{Sys.Date()}.rds")
)
saveRDS(
  ref_coeffs_r,
  glue("{wkdir}/data/interim/osNMF/ABCA1&2osNMF_coeffs_K11_{Sys.Date()}_.rds")
)

```

## Plotting ABCA NMF DIMS

```{r, eval=FALSE}
# adding NMF results to Seurat object
ref_seur_trim <- AddMetaData(
  object = ref_seur_trim,
  metadata = ref_PPs_r
)
ref_seur_trim$slice_id <- ref_seur_trim$brain_section_label.x
ref_seur_trim@meta.data %>% glimpse

nmdf_dims <- ref_seur_trim@meta.data %>% 
  colnames() %>% 
  keep(grepl("osNMF", .))

for (d in paste0("osNMF_", 1:length(nmdf_dims))){
  message(glue("Plotting: {d}"))
  p <- ref_seur_trim@meta.data %>%
    rownames_to_column(var = "cell_id") %>%
    arrange(z, !!sym(d)) %>%
    mutate(cell_id, factor(cell_id)) %>%
    mutate(slice_id, factor(slice_id)) %>%
    ggplot(aes(x, y, color = !!sym(d))) +
    geom_point(size = 0.3) +
    scale_color_viridis() +
    coord_fixed() +
    scale_y_reverse() +
    facet_wrap(~slice_id, ncol = 9) +
    theme_bw()

  ggsave(
    glue("{wkdir}/figures/osNMF_K11_ABCA/K11_{d}_{Sys.Date()}.png"),
    p,
    width = 25, height = 20
  )
}

```


# Re-running staNMF - post Quality Control (after running notebook 04_QC-cell-filtering.qmd)

Prepping R objects for python cells

```{r, eval=FALSE}
merged_rois <- readRDS(
    glue("{wkdir}/data/interim/merged_roi_seurat_filtered_2024-07-22.rds")
)

seurat_data <- merged_rois %>% 
  GetAssayData(layer = "data")
filtered_data <- as.matrix(seurat_data)  # Convert to matrix

```

Running staNMF on all cell types - using a prespecified k (11)

```{python}
import numpy as np
from sklearn.decomposition import NMF as sklearn_nmf
import sys
sys.path.append('/central/groups/mthomson/jboktor/spatial_genomics/osNMF/osNMF')
from knn_impute import kNN_imputation
from morans_index import MoranIforPPs
from other_functions import show_graph_with_labels, filter_genes, weighted_correlation
from sklearn_nmf import sklearn_nmf
from stability import findcorrelation, HungarianError, amariMaxError, instability

count_mat = r.filtered_data
nmf = sklearn_nmf(n_components=11, l1_ratio=1, alpha=0, random_state=1, max_iter=1000)
nmf.fit(np.maximum(count_mat, 0))

PPs = nmf.components_
coeffs = nmf.transform(np.maximum(count_mat, 0))

# convert into R objects
r.PPs = PPs
r.coeffs = coeffs
```

Convert the results back into R objects

```{r}
PPs_r <- matrix(PPs, nrow = dim(PPs)[1], ncol = dim(PPs)[2]) %>%
  t() %>%
  set_colnames(paste0("osNMF_", 1:dim(PPs)[1]))
coeffs_r <- matrix(coeffs, nrow = dim(coeffs)[1], ncol = dim(coeffs)[2])

cat("PPs shape:", dim(PPs_r), "\n")
cat("Coeffs shape:", dim(coeffs_r), "\n")

saveRDS(
  PPs_r,
  glue("{wkdir}/data/interim/osNMF/osNMF_PPs_K11_postqc_{Sys.Date()}.rds")
)
saveRDS(
  coeffs_r,
  glue("{wkdir}/data/interim/osNMF/osNMF_coeffs_K11_postqc_{Sys.Date()}_.rds")
)

```

## Plotting seqFISH NMF DIMS

```{r, eval=FALSE}
# fig dir
nmf_figs <- glue("{wkdir}/figures/osNMF_K11_postqc/{Sys.Date()}/")
dir.create(nmf_figs, showWarnings = FALSE, recursive = TRUE)
dir.create(glue("{nmf_figs}/scatterplots"), 
  showWarnings = FALSE, recursive = TRUE
  )

# adding NMF results to Seurat object
merged_rois <- AddMetaData(
  object = merged_rois,
  metadata = PPs_r
)
merged_rois@meta.data %>% glimpse

nmdf_dims <- merged_rois@meta.data %>% 
  colnames() %>% 
  keep(grepl("osNMF_\\d+_postqc$", .))

for (d in nmdf_dims ){  
  message(glue("Plotting: {d}"))
  p <- merged_rois@meta.data %>%
    rownames_to_column(var = "cell_id") %>%
    arrange(!!sym(d)) %>%
    mutate(cell_id, factor(cell_id)) %>%
    arrange(group) %>%
    mutate(slice_id, factor(slice_id)) %>%
    ggplot(aes(center_x, center_y, color = !!sym(d))) +
    geom_point(size = 0.2) +
    scale_color_viridis() +
    coord_fixed() +
    scale_y_reverse() +
    facet_wrap(~slice_id, ncol = 4) +
    theme_bw()

  ggsave(
    glue("{nmf_figs}/scatterplots/K11_{d}.png"),
    p,
    width = 20, height = 20
  )
}

```


Save updated osNMF data into Seurat object

```{r, eval = FALSE}
# Adding staNMF data to object ---
merged_rois <- readRDS(
    glue("{wkdir}/data/interim/merged_roi_seurat_filtered_2024-07-22.rds")
)
PPs_r <- readRDS(
  glue("{wkdir}/data/interim/osNMF/osNMF_PPs_K11_postqc_2024-07-24.rds")
)
colnames(PPs_r) <- paste0(colnames(PPs_r), "_postqc")
merged_rois <- Seurat::AddMetaData(
  object = merged_rois,
  metadata = PPs_r
)

saveRDS(
  merged_rois,
    glue("{wkdir}/data/interim/merged_roi_seurat_filtered_staNMF-updated_{Sys.Date()}.rds")
)
```


# osNMF Analysis with cell annotations
Plotting osNMF cell type composition across dims

```{r, eval = FALSE}
nmf_figs <- glue("{wkdir}/figures/osNMF_K11_postqc/2024-07-24")
nmf_cell_type_df %>% glimpse
meta_df <- merged_rois@meta.data %>% glimpse()

quantile_cols <- colnames(meta_df) %>% 
 keep(~ grepl("osNMF_\\d+_postqc_Q\\d+", .))

meta_df %<>%
  mutate_at(vars(all_of(quantile_cols)), 
    ~ gsub("Q", "", .) %>% as.numeric ) %>%
  glimpse()

nmf_cell_type_df <- tibble()
for (qc in quantile_cols) {
  df <- meta_df %>%
    filter(.data[[qc]] <= 3) %>%
    group_by( n   n) %>%
    dplyr::summarize(n = n()) %>%
    mutate(nmf_dim = qc) %>%
    glimpse()
  nmf_cell_type_df %<>% bind_rows(df)
}


```


Isolating the top 5% of cells with highest NMF scores
```{r, eval = FALSE}

p_nmf_heatmap <- nmf_cell_type_df %>%
  arrange(nmf_dim) %>%
  mutate(nmf_dim = factor(nmf_dim, levels = unique(nmf_dim))) %>%
  ggplot(aes(y = singleR_labels, x = nmf_dim)) +
  geom_point(aes(size = n)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(
  glue("{nmf_figs}/nmf_cell_type_heatmap_{Sys.Date()}.png"),
  p_nmf_heatmap,
  width = 7, height = 7
)


# Create a clustering function
cluster_nmf_data <- function(df) {
  require(seriation)
  # Create a matrix for seriation
  matrix_data <- df %>%
    pivot_wider(
      names_from = nmf_dim,
      values_from = n,
      values_fill = list(n = 0)
    ) %>%
    column_to_rownames("singleR_labels") %>%
    as.matrix()

  # Perform seriation
  seriated_cells <- seriate(matrix_data, method = "PCA")
  seriated_nmf_dim <- seriate(t(matrix_data), method = "PCA")

  # Get the new order
  new_order_cells <- seriation::get_order(seriated_cells)
  new_order_nmf_dims <- seriation::get_order(seriated_nmf_dim)

  # Reorder the data frame
  df <- df %>%
    mutate(
      singleR_labels = factor(singleR_labels,
        levels = rownames(matrix_data)[new_order_cells]
      ),
      nmf_dim = factor(nmf_dim,
        levels = colnames(matrix_data)[new_order_nmf_dims]
      )
    )

  return(df)
}

nmf_cell_type_df_clust <- cluster_nmf_data(nmf_cell_type_df)

p_nmf_bars <- nmf_cell_type_df_clust %>%
  ggplot(aes(x = nmf_dim, y = n, fill = singleR_labels)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = abca_color_pal[["cluster_colors"]]) +
  theme_bw() +
  labs(x = NULL, y="Cell type composition of top 30th percentile of NMF coefficient") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(
  glue("{nmf_figs}/nmf_cell_type_bars_{Sys.Date()}.png"),
  p_nmf_bars,
  width = 7, height = 6
)

```

![osNMF cell type composition heatmap](../figures/osNMF_K11_postqc/2024-07-24/nmf_cell_type_bars_2024-07-24.png)


### NMF DEG Analysis

```{r, eval = FALSE}

# merged_rois <- readRDS(
#   glue(
#     "{wkdir}/data/interim/",
#     "2024-06-10_merged_roi_seurat_singleR-annot_osNMF.rds"
#   )
# )
merged_rois@meta.data %>% glimpse

nmf_vars <- merged_rois@meta.data %>%
  colnames() %>%
  grep("osNMF", ., value = TRUE)


nmf_deg_df <- tibble()
for (nmf_col in nmf_vars) {
  message(glue("Processing: {nmf_col}"))
  df <- merged_rois@meta.data %>%
    dplyr::select(singleR_labels, nmf_var = !!sym(nmf_col))

  quantiles <- quantile(df$nmf_var, probs = seq(0, 1, by = 0.05)) %>%
    unique()

  df <- df %>%
    mutate(quantile = cut(
      nmf_var,
      breaks = quantiles,
      labels = rev(paste0("Q", seq(1, length(quantiles) - 1))),
      include.lowest = TRUE,
      right = FALSE
    ))
  
  # adding metadat to seurat object
  merged_rois$nmf_quantiles <- df$quantile
  
  # select cells that are in the top 5% quantile
  merged_rois_subset <- subset(merged_rois, nmf_quantiles == "Q1")

  # perform marker gene analysis across SPF/WildR for NMF cells
  deg_res <- Seurat::FindMarkers(
    merged_rois_subset,
    ident.1 = glue("wildr"),
    group.by = "group"
  ) %>%
    as.data.frame() %>%
      rownames_to_column(var = "gene_symbol") %>%
      mutate(nmf_dimension = nmf_col)

    nmf_deg_df %<>% bind_rows(deg_res)
}
nmf_deg_df %>% glimpse

saveRDS(
  nmf_deg_df,
  glue(
    "{wkdir}/data/interim/osNMF/",
    "NMF_DEG_df_{Sys.Date()}.rds"
  )
)

```

Visualizing NMF DEG results

```{r, eval = FALSE}
nmf_deg_df <- readRDS(
  glue(
    "{wkdir}/data/interim/osNMF/",
    "NMF_DEG_df_2024-06-24.rds"
  )
)

nmf_order <- nmf_deg_df %>%
  mutate(avg_log2FC = if_else(avg_log2FC > 0, avg_log2FC, 0)) %>%
  group_by(nmf_dimension) %>%
  summarize(sum_log2FC = sum(avg_log2FC)) %>%
  arrange(sum_log2FC) %>%
  pull(nmf_dimension)


p_nmf_deg_barplot <- nmf_deg_df %>%
  filter(p_val_adj <= 0.05) %>%
  mutate(nmf_dimension = factor(nmf_dimension, levels = nmf_order)) %>%
    ggplot(aes(
      x = avg_log2FC,
      y = nmf_dimension
    )) +
    geom_bar(
      stat = "identity", 
      position = "stack",
      width = 0.6,
      fill = "white",
      aes(color = gene_symbol)
    ) +
      labs(y = NULL, x = expression("Cumulative" ~ log[2] * "FC WildR/SPF")) +
      theme_bw() +
      scale_color_viridis_d(option = "turbo") +
      theme(legend.position = "none")
    
ggsave(
  glue("{wkdir}/figures/osNMF_K11/NMF_DEG_barplot_{Sys.Date()}.png"),
  p_nmf_deg_barplot,
  width = 6, height = 4
)

```


GO enrichment analysis on NMF DEG results

```{r, eval = FALSE}
nmf_deg_df <- readRDS(
  glue(
    "{wkdir}/data/interim/osNMF/",
    "NMF_DEG_df_2024-06-24.rds"
  )
)

gene_symbols <- nmf_deg_df$gene_symbol %>% unique()
gene_symbol_map <- data.frame(
  gene_symbol = gene_symbols,
  gene_symbol_formatted = tools::toTitleCase(tolower(gene_symbols))
)

deg_df <- nmf_deg_df %<>%
  left_join(gene_symbol_map, by = "gene_symbol") %>%
  mutate(direction = if_else(
    avg_log2FC > 0, "WildR Upregulated", "WildR Downregulated"
  ))

deg_df %>% glimpse

# format params for GO enrichment
params_1 <- expand.grid(
  unique(deg_df$nmf_dimension),
  c( "WildR Upregulated", "WildR Downregulated")
) %>% 
dplyr::rename(
  nmf_dimension = Var1,
  direction = Var2
)
params_2 <- expand.grid(
  c("BP", "CC", "MF"),
  c( "WildR Upregulated", "WildR Downregulated")
) %>% 
dplyr::rename(
  Ontology = Var1,
  direction = Var2
)

go_params <- params_1 %>% 
  full_join(params_2, by = "direction", relationship = "many-to-many") %>% 
  glimpse()

future::plan("multisession", workers = 64)
enrichGO_res <-
  furrr::future_pmap(
    list(
      go_params$Ontology,
      go_params$direction,
      go_params$nmf_dimension
    ),
    ~ enrichGO(
      gene = deg_df %>%
        filter(direction == ..2 & nmf_dimension == ..3) %>%
        pull(gene_symbol_formatted),
      OrgDb = org.Mm.eg.db,
      keyType = "SYMBOL",
      ont = ..1,
      pAdjustMethod = "BH",
      pvalueCutoff = 0.05,
      qvalueCutoff = 0.2,
      readable = TRUE
    ) %>%
      as.data.frame() %>%
      mutate(Ontology = ..1, direction = ..2, nmf_dimension = ..3)
  ) %>%
  bind_rows() %>%
  glimpse()

saveRDS(
  enrichGO_res,
  glue(
    "{wkdir}/data/interim/osNMF_K11",
    "NMF_DEG_enrichGO_res_{Sys.Date()}.rds"
  )
)


enrichGO_res <- readRDS(
  glue(
    "{wkdir}/data/interim/osNMF_K11",
    "NMF_DEG_enrichGO_res_2024-06-25.rds"
  )
)
enrichGO_res %>% glimpse
 
for (direc in c("WildR Upregulated", "WildR Downregulated")) {
  for (cluster in unique(enrichGO_res$nmf_dimension)) {
    message(glue("Plotting: {cluster} - {direc}"))
    p <- enrichGO_res %>%
      filter(direction == direc) %>%
      filter(nmf_dimension == cluster) %>%
      filter(qvalue <= 0.05) %>%
      slice_min(qvalue, n = 30) %>%
      slice_max(Count, n = 20, with_ties = FALSE) %>%
      mutate(Description = glue("{Description}:{Ontology}")) %>% 
      ggplot(aes(x = Count, y = fct_reorder(Description, Count))) +
      geom_col(width = 0.4, fill = "red") +
      theme_bw() +
      labs(
        x = "Gene Count", y = NULL,
        title = glue("{cluster}: {direc}")
      )

    ggsave(
      glue(
        "{wkdir}/figures/osNMF_K11/GO/",
        "GO_enrichment_{cluster}_{direc}_{Sys.Date()}.png"
        ),
      p,
      width = 8, height = 5.5
    )
  }
}

deg_df %>%
  filter(nmf_dimension == "osNMF_8") %>%
  arrange(desc(avg_log2FC)) %>%
  head()

```


Visualizing expression of serial genes across slices to identify issues

```{r, eval = FALSE}
# Function to plot gene epxression on spatial data
plot_xy_gene_exp <- function(seur, gene) {
  gene = toupper(gene)
  meta_df <- merged_rois@meta.data
  meta_df <- bind_cols(meta_df, 
    genename = merged_rois@assays$SCT@counts[gene, ]) %>% 
    dplyr::mutate(slice_id = 
      glue("{toupper(group)} {run} {roi}")) %>%
    arrange(group, run, genename)
    
  p <- meta_df %>%
    ggplot(aes(x = center_x, y = - center_y)) +
    geom_point(aes(color = genename), size = 0.2) +
    facet_wrap(~ slice_id, scale = "free", nrow = 4) +
    labs(color = gene) +
    scale_color_viridis_c(option = "magma") +
    theme_bw()
  return(p)
}

neuromapper_serial_genes <- read.csv(
  glue("{wkdir}/data/input/NeuroMapper_Channel2_gene_list.csv"),
  row.names = NULL
)
serial_genes <- toupper(neuromapper_serial_genes$gene_name)

for (gid in neuromapper_serial_genes$gene_name){
  message(glue("Plotting: {gid}"))
  figout <- glue("{wkdir}/figures/gene_exp_scatter/{gid}_expression_{Sys.Date()}.png")
  if (file.exists(figout)) next
  
  p <- plot_xy_gene_exp(merged_rois, gid)
  ggsave(
    filename = figout,
    p,
    width = 22, height = 20
  )
}

```



Examining gene coeff. for each NMF dimension to identify which genes are contributing to artifacts. 

```{r, eval=FALSE}

# read in gene coef from staNMF
nmf_coef <- readRDS(
  glue("{wkdir}/data/interim/osNMF/osNMF_coeffs_K11_2024-07-15_.rds")
)

nmf_coef %>% glimpse

rownames(nmf_coef) <- rownames(merged_rois)
coef_df <- nmf_coef %>% 
  as.data.frame() %>%
  rownames_to_column(var = "gene") %>%
  pivot_longer(!gene, 
  names_to = "nmf_dim", 
  values_to = "coef") %>%
  mutate(serial = gene %in% serial_genes) %>%
  mutate(nmf_dim = gsub("V", "osNMF", nmf_dim)) %>%
  glimpse()

# serial genes contribute at a much higher level than other genes
p_nmf_coef_heat <- coef_df %>% 
  # filter(gene %in% serial_genes) %>%
  ggplot(aes(x = nmf_dim, y = gene)) +
  geom_point(aes(color = coef, size = coef)) +
  theme_bw() +
  scale_color_viridis_c(option = "magma") +
  facet_grid(rows = vars(serial), scales = "free", space = "free") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

ggsave(
  filename = glue("{wkdir}/figures/osNMF_K11/nmf_coef_heatmap_{Sys.Date()}.png"),
  p_nmf_coef_heat,
  width = 8, height = 15
)

# But what about when matching genes at similar exp levels?
# Calculate row-wise CV
cv <- function(x) {
  sd(x) / mean(x)
}
rowwise_cv <- apply(merged_rois@assays$SCT@counts, 1, cv)
rowwise_mean <- apply(merged_rois@assays$SCT@counts, 1, mean)
rowwise_median <- apply(merged_rois@assays$SCT@counts, 1, median)

gene_stats_df <- data.table(
  gene = names(rowwise_cv),
  cv = unname(rowwise_cv),
  mean = unname(rowwise_mean),
  median = unname(rowwise_median)
) %>% glimpse()

# visualizing mean NMF coef contr. by gene by CV
cv_nmf_coef <- coef_df %>% 
  group_by(gene) %>%
  summarize(mean_coef = mean(coef)) %>%
  left_join(gene_stats_df, by = "gene") %>%
  mutate(serial = gene %in% serial_genes)

p_cv_by_coef <- cv_nmf_coef %>%
  ggplot(aes(x = cv, y = mean_coef)) +
  geom_point(aes(color = serial)) +
  scale_color_manual(values = c("grey", "red")) +
  theme_bw()
p_mean_by_coef <- cv_nmf_coef %>%
  ggplot(aes(x = mean, y = mean_coef)) +
  geom_point(aes(color = serial)) +
  scale_color_manual(values = c("grey", "red")) +
  theme_bw()
p_median_by_coef <- cv_nmf_coef %>%
  ggplot(aes(x = median, y = mean_coef)) +
  geom_point(aes(color = serial)) +
  scale_color_manual(values = c("grey", "red")) +
  theme_bw()

ggsave(
  filename = glue("{wkdir}/figures/osNMF_K11/cv_by_coef_{Sys.Date()}.png"),
  p_cv_by_coef,
  width = 8, height = 6
)
ggsave(
  filename = glue("{wkdir}/figures/osNMF_K11/mean_by_coef_{Sys.Date()}.png"),
  p_mean_by_coef,
  width = 8, height = 6
)
ggsave(
  filename = glue("{wkdir}/figures/osNMF_K11/median_by_coef_{Sys.Date()}.png"),
  p_median_by_coef,
  width = 8, height = 6
)

gene_order <- coef_df %>%
  group_by(gene) %>%
  summarize(mean_coef = mean(coef)) %>%
  arrange(desc(mean_coef)) %>%
  pull(gene)

p_nmf_coef_dp <- coef_df %>% 
  mutate(gene = factor(gene, levels = gene_order)) %>%
  ggplot(aes(x = gene, y = coef)) +
  geom_point(aes(color = serial)) +
  theme_bw() +
  scale_color_manual(values = c("grey", "red")) +
  labs(x = NULL, y = "NMF Coefficient") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

ggsave(
  filename = glue("{wkdir}/figures/osNMF_K11/nmf_coef_dotplot_{Sys.Date()}.png"),
  p_nmf_coef_dp,
  width = 15, height = 6
)

# What are the genes highly expressed in NMF2?
coef_df %>% 
  filter(nmf_dim == "osNMF1") %>%
  arrange(desc(coef)) %>%
  head(30) %>% 
  print(n=30)

```

A tibble: 30 Ã— 4
   gene     nmf_dim  coef serial
   <chr>    <chr>   <dbl> <lgl> 
 1 CAMK2N1  osNMF1  18.6  TRUE  
 2 CALM2    osNMF1  15.4  TRUE  
 3 SLC1A2   osNMF1  14.9  TRUE  
 4 SLC17A6  osNMF1  14.3  TRUE  
 5 APP      osNMF1  12.1  TRUE  
 6 GAP43    osNMF1  12.1  TRUE  
 7 APOE     osNMF1  11.9  TRUE  
 8 CALM1    osNMF1  11.9  TRUE  
 9 CALM3    osNMF1  11.5  TRUE  
10 SNAP25   osNMF1  11.2  TRUE  
11 ATP1A3   osNMF1  10.9  TRUE  
12 MEF2C    osNMF1  10.5  TRUE  
13 CCK      osNMF1  10.3  TRUE  
14 MAPT     osNMF1  10.1  TRUE  
15 SERPINE2 osNMF1   7.68 TRUE  
16 NRGN     osNMF1   7.29 TRUE  
17 CRYAB    osNMF1   6.51 TRUE  
18 SLC17A7  osNMF1   6.17 TRUE  
19 SST      osNMF1   5.41 TRUE  
20 SLC1A3   osNMF1   4.96 TRUE  
21 NPY      osNMF1   3.62 TRUE  
22 KIF5A    osNMF1   3.58 FALSE 
23 VIP      osNMF1   3.25 TRUE  
24 SLC17A8  osNMF1   3.10 TRUE  
25 C1QA     osNMF1   2.65 TRUE  
26 FOS      osNMF1   2.52 TRUE  
27 PVALB    osNMF1   2.48 TRUE  
28 C1QB     osNMF1   2.11 TRUE  
29 LINGO1   osNMF1   1.82 FALSE 
30 CAMK2A   osNMF1   1.78 FALSE

