---
title: "WildR Brain Tissue Spatial Transcriptomics Analysis"
editor: source
author: "Joe Boktor"
date: '2024-04-13'
format: 
  html:
    font-family: helvetica neue
    page-layout: full
    toc: true
    toc-location: left
    toc-depth: 5
    self-contained: true
    code-fold: false
    code-tools: true
    fig-align: center
    grid:
      sidebar-width: 200px
      body-width: 1100px
      margin-width: 200px
      gutter-width: 1.5em
---

# Background

Here we analyze

# Analysis Setup

Environment setup.

```{r}
#| warning: false

library(tidyverse)
library(magrittr)
library(glue)
library(Seurat)
library(grid)
library(biomaRt)
library(SingleR)
library(batchtools)
# stats 
library(lme4)
library(broom)
# parallelization
library(BiocParallel)
library(future)
library(furrr)
# plotting
library(RColorBrewer)
library(ggsci)
library(Matrix)
library(plotly)
library(DT)
library(gtsummary)
library(aplot)
library(patchwork)
library(viridis)

# setting paths
homedir <- "/central/groups/mthomson/jboktor"
wkdir <- glue("{homedir}/spatial_genomics/jess_2024-01-23")
source(glue("{wkdir}/notebooks/R_scripts/_misc_functions.R"))

# 12gb  limit (1500*1024^2 = 1572864000) * 8
options(future.globals.maxSize= 12582912000)
future::plan("multisession", workers = 8)

```


### Summary stats of data

```{r, eval = FALSE}
abca_color_pal <- readRDS(glue("{wkdir}/data/interim/abca_color_pal.rds"))
merged_rois <- readRDS(
    glue(
      "{wkdir}/data/interim/",
      "merged_roi_seurat_filtered_staNMF-updated_2024-07-24.rds"
    )
)
meta_df <- merged_rois@meta.data %>% glimpse()

# saveRDS(
#   meta_df,
#     glue(
#       "{wkdir}/data/interim/",
#       "merged_roi_seurat_filtered_staNMF-updated_2024-07-24-METADATA.rds"
#     )
# )


```

Plotting UMAP clusters

```{r, eval = FALSE}
p_umap_clust <- DimPlot(merged_rois,
  group.by = "seurat_clusters", 
  label = TRUE, repel = TRUE
)

p_umap_celltype <- DimPlot(merged_rois,
  group.by = "singleR_labels",
  label = TRUE, 
  label.box = TRUE,
  repel = TRUE
) +
    coord_fixed() +
  scale_color_manual(values = abca_color_pal[["cluster_colors"]]) +
  scale_fill_manual(values = abca_color_pal[["cluster_colors"]])


ggsave(
  glue("{wkdir}/figures/umaps/umap_seurat_clusters_{Sys.Date()}.png"),
  p_umap_clust,
  width = 9, height = 7
)

ggsave(
  glue("{wkdir}/figures/umaps/umap_singler_celltypes_{Sys.Date()}.png"),
  p_umap_celltype,
  width = 15, height = 10
)

```


```{r, eval = FALSE}
tab <- table(
  Assigned = meta_df$singleR_labels,
  Clusters = meta_df$seurat_clusters
)
p_cluster_annot_heatmap <-
  pheatmap::pheatmap(log10(tab + 10),
  # pheatmap::pheatmap(tab,
    color = colorRampPalette(c("white", "blue"))(10)
  )
ggsave(
  glue("{wkdir}/figures/SingleR/",
  "SingleR_seurat-cluster-annotation-heatmap_{Sys.Date()}.png"
  ),
  p_cluster_annot_heatmap,
  width = 9, height = 7
)

```


![Heatmap showing cell membership of unsupervised clusters to SingleR cell type annotations. Color indicates Log10 counts of cells per grouping.](../figures/SingleR/SingleR_seurat-cluster-annotation-heatmap_2024-07-22.png){#fig-unsuper-clust-vs-singleR}


![UMAP showing unsupervised cluster detection annotated with SingleR cell types](../figures/umaps/umap_singler_celltypes_2024-07-22.png){#fig-umap-singleR}


Plotting gene expression across cell types


```{r, eval = FALSE}
p <- DoHeatmap(
  merged_rois,
  features = VariableFeatures(merged_rois)[1:250],
  # cells = 1:30000, 
  cells = sample(1:ncol(merged_rois), 200000), 
  size = 4,
  group.by = "singleR_labels",
  angle = 90,
  raster = FALSE
) +
  NoLegend()

ggsave(
  glue("{wkdir}/figures/celltype_plots/celltype-heatmap_{Sys.Date()}.png"),
  p,
  width = 40, height = 30,
  dpi = 600
)

```

![Cell type marker heatmap](../figures/celltype_plots/celltype-heatmap_2024-07-22.png){#fig-celltype-marker-heatmap}



Plotting brain slices with cell type annotations

```{r, eval = FALSE}

slice_view_dir <- glue("{wkdir}/figures/SliceView/{Sys.Date()}")
dir.create(slice_view_dir, recursive = TRUE, showWarnings = FALSE)

for (cell_label in unique(meta_df$singleR_labels)) {
  message(glue("Plotting {cell_label}"))
  p_brain <- meta_df %>%
    mutate(clust_highlight = case_when(
      singleR_labels == cell_label ~ TRUE,
      TRUE ~ FALSE
    )) %>%
    arrange(slice_id, group, run) %>%
    mutate(slice_id = factor(slice_id)) %>%
    arrange(clust_highlight) %>%
    # mutate(seurat_clusters = factor(seurat_clusters)) %>%
    ggplot(aes(x = center_x, y = center_y, color = clust_highlight)) +
    geom_point(size = 0.05, aes(alpha = clust_highlight)) +
    scale_color_manual(values = c("TRUE" = "red", "FALSE" = "grey")) +
    scale_alpha_manual(values = c("TRUE" = 0.3, "FALSE" = 0.1)) +
    theme_bw() +
    coord_fixed() +
    scale_y_reverse() +
    labs(title = cell_label, x = NULL, y = NULL) +
    facet_wrap(~slice_id, ncol = 4) +
    theme(legend.position = "none")

  ggsave(p_brain,
    file = glue("{slice_view_dir}/{cell_label}_plot_{Sys.Date()}.png"),
    width = 16, height = 16
  )
}

p_brain_all <- meta_df %>%
      arrange(slice_id) %>%
      mutate(slice_id = factor(slice_id)) %>%
      ggplot(aes(x = center_x, y = center_y, color = singleR_labels)) +
      geom_point(size = 0.1, alpha = 0.7) +
      scale_color_manual(values = abca_color_pal[["class_colors"]]) +
      theme_bw() +
      coord_fixed() +
      scale_y_reverse() +
      labs(x = NULL, y = NULL) +
      guides(colour = guide_legend(override.aes = list(size=3))) +
      facet_wrap(~slice_id, ncol = 4)

ggsave(p_brain_all,
    file = glue("{wkdir}/figures/SliceView/00_ALL-CELLS_v2_{Sys.Date()}.png"),
    width = 20, height = 20
  )

```

![Plots of all cells and slices post QC](../figures/SliceView/00_ALL-CELLS_v2_2024-07-22.png){#fig-all-cells}

Plotting cell type proportions across slices.

```{r, eval = FALSE}

count_df <- table(
  merged_rois$group,
  merged_rois$slice_id,
  merged_rois$singleR_labels
  ) %>% 
  as.data.frame() %>%
  dplyr::rename(
    group = Var1,
    slice_id = Var2,
    cell_label = Var3,
    counts = Freq
    ) %>% 
  filter(counts > 0) %>% 
  group_by(slice_id) %>% 
  dplyr::mutate(relab = counts/sum(counts) * 100)

# count_df %>% glimpse

p_bar <- count_df %>% 
  ggplot(aes(x=slice_id, y=relab, fill = cell_label)) +
  geom_bar(stat = "identity", color = "black", linewidth = 0.3) +
  scale_fill_manual(values = abca_color_pal[["cluster_colors"]]) +
  facet_wrap(~group, ncol = 2, scales = "free") +
  labs(x =NULL, y= "Relative Abundance (%)", fill = NULL) +
  theme_classic()  +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(
  glue("{wkdir}/figures/celltype_plots/celltype_barplot_{Sys.Date()}.png"),
  p_bar,
  width = 9, height = 5
)

p_bar_counts <- count_df %>% 
  ggplot(aes(x=slice_id, y=counts, fill = cell_label)) +
  geom_bar(stat = "identity", color = "black", linewidth = 0.3) +
  scale_fill_manual(values = abca_color_pal[["cluster_colors"]]) +
  facet_wrap(~group, ncol = 2, scales = "free_x") +
  labs(x =NULL, y= "Cell Counts", fill = NULL) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(
  glue("{wkdir}/figures/celltype_plots/celltype_barplot_counts_{Sys.Date()}.png"),
  p_bar_counts,
  width = 9, height = 5
)

```



![](../figures/celltype_plots/celltype_barplot_counts_2024-07-22.png){#fig-bar-counts}


![](../figures/celltype_plots/celltype_barplot_2024-07-22.png){#fig-bar-relab}


Linear model formula for cell type counts and proportion analysis across slices, accouting for variability in bregma as a fixed effect.

$\text{counts} \sim \text{group} + \text{bregma} + \text{Run Date}$ 

```{r, eval = FALSE}

core_meta_df <- merged_rois@meta.data %>%
  select(run, roi, group, bregma, slice_id) %>%
  mutate(slice_id = factor(slice_id)) %>%
  distinct() %>%
  glimpse()

count_df_joined <- count_df %>%
  left_join(core_meta_df)

lm_results_counts <- count_df_joined %>%
  split(.$cell_label) %>%
  purrr::map_df(~ {
    model <- lm(counts ~ group + bregma + run, data = .x)
    tidy(model, effects = "fixed") %>%
      mutate(cell_label = unique(.x$cell_label))
  })

lm_results_relab <- count_df_joined %>%
  split(.$cell_label) %>%
  purrr::map_df(~ {
    model <- lm(relab ~ group + bregma + run, data = .x)
    tidy(model, effects = "fixed") %>%
      mutate(cell_label = unique(.x$cell_label))
  })

# View the results
lm_results_counts %>% glimpse

p_lm_counts <- lm_results_counts %>%
  mutate(signif = p.value <= 0.05) %>%
  filter(term == "groupwildr") %>%
  # filter(term != "(Intercept)") %>%
  ggplot(aes(
    y = fct_reorder(cell_label, estimate),
    x = estimate,
  )) +
  # scale_x_log10() +
  geom_bar(stat = "identity", width = 0.5, aes(fill = signif)) +
    theme_bw() +
    scale_fill_manual(values = c("TRUE" = "red", "FALSE" = "grey")) +
    labs(
      title = "Difference in Counts by Cell Label",
      y = NULL,
      x = "Beta Coefficient WildR / SPF "
    ) +
    facet_wrap(~term, nrow = 1) +
      theme(legend.position = "top")

p_lm_relab <- lm_results_relab %>%
  mutate(signif = p.value <= 0.05) %>%
  filter(term == "groupwildr") %>%
  # filter(term != "(Intercept)") %>%
  ggplot(aes(
    y = fct_reorder(cell_label, estimate),
    x = estimate,
  )) +
  # scale_x_log10() +
  geom_bar(stat = "identity", width = 0.5, aes(fill = signif)) +
    theme_bw() +
    scale_fill_manual(values = c("TRUE" = "red", "FALSE" = "grey")) +
    labs(
      title = "Difference in Counts by Cell Label",
      y = NULL,
      x = "Beta Coefficient WildR / SPF "
    ) +
    facet_wrap(~term, nrow = 1) +
      theme(legend.position = "top")


ggsave(
  glue("{wkdir}/figures/Celltype_prop_LM/",
    "celltype_counts_lm_results-counts_{Sys.Date()}.png"),
  p_lm_counts,
  width = 5, height = 6
)
ggsave(
  glue("{wkdir}/figures/Celltype_prop_LM/",
    "celltype_counts_lm_results-relab_{Sys.Date()}.png"),
  p_lm_relab,
  width = 5, height = 6
)

```


![](../figures/Celltype_prop_LM/celltype_counts_lm_results-relab_2024-07-22.png){#fig-lm-relab}

![](../figures/Celltype_prop_LM/celltype_counts_lm_results-counts_2024-07-22.png){#fig-lm-counts}


Plotting specific cell type prop box plots

```{r, eval = FALSE}

for (type in unique(count_df_joined$cell_label)) {
  p <- count_df_joined %>%
    filter(cell_label == type) %>%
    mutate(group = toupper(group)) %>%
    ggplot(aes(x = group, y = relab)) +
    geom_boxplot(outlier.alpha = 0) +
    geom_point(aes(color = bregma), 
      size = 3, alpha = 0.8,
      position = position_jitter(width = 0.2)) +
    labs(x = NULL, y = "Relative Proportion (%)", 
      title = type) +
    scale_color_viridis(option = "plasma") +
    theme_classic()

  ggsave(
    glue("{wkdir}/figures/Celltype_prop_boxplots/{type}_{Sys.Date()}.png"),
    p,
    width = 2.5, height = 3
  )  
}

```


![](../figures/Celltype_prop_boxplots/12 HY GABA_2024-07-22.png){#fig-12hygaba-relab}


```{r}
sessionInfo()
```