---
title: "Differential Expression Analysis Visualization and Interpretation"
editor: source
author: "Joe Boktor"
date: '2026-01-01'
format: 
  html:
    self-contained: true
    code-fold: true
    code-tools: true
    fig-align: right
    fig-height: 8
    fig-width: 14 
---


```{r}
library(Seurat)
library(Matrix)
library(irlba)
library(scico)
library(tidyverse)
library(patchwork)
library(glue)
library(presto)
library(hdf5r)
library(tidyseurat)
library(Augur)
library(batchtools)

library(tictoc)
library(future)
library(strex)
library(cowplot)
library(aplot)
library(SingleCellExperiment)
library(BiocParallel)
library(ggside)

theme_set(theme_bw())

home_dir <- "/resnick/groups/MazmanianLab/jboktor"
rnk_dir <- "/resnick/groups/mthomson/jboktor"
split_pipe_dir <- glue("{rnk_dir}/split-pipe_workflow")
wkdir <- glue("{home_dir}/WILDRxSPF_brains")
anndata_path <- glue("{wkdir}/data/input/parse_bio_anndata")
data_path <- glue("{wkdir}/data/interim")
fig_path <- glue("{wkdir}/figures")
abca_color_pal <- readRDS(glue("{wkdir}/data/interim/abca_color_pal.rds"))

source(glue("{wkdir}/notebooks/R_scripts/_misc_functions.R"))

# Set memory limit for future
options(future.globals.maxSize = 200000 * 1024^2)  # Set to ~200GB
# Use more memory-efficient serialization
options(future.serialization.progressr = FALSE)
options(future.serialization.progressr.enable = FALSE)

# future::plan(multisession, workers = 3)
# future::plan(sequential)

```


Loading Seurat object


```{r}
seurat_obj <- readRDS( 
    glue("{data_path}/seurat/snRNASeq/",
    "seurat_obj_onlyparsefilters_celltyped_2025-12-27.rds")
    )
seurat_obj
# Subset the seurat object to only keep cells that pass the QC filters
seurat_obj <- seurat_obj %>% subset(subset = keep_cell)
seurat_obj@meta.data %>% glimpse()

allen_annot_df <- seurat_obj@meta.data %>% 
  select(class_name, class_label, subclass_name, 
    subclass_label, supertype_name, supertype_label) %>% 
  distinct() %>% 
  glimpse()

allen_annot_long <- allen_annot_df %>% 
  pivot_longer(
    cols = everything(),
    names_to = c("cell_type_level", ".value"),
    names_pattern = "(class|subclass|supertype)_(name|label)"
  ) %>% 
  dplyr::rename(cell_type_name = name, cell_type_label = label) %>% 
  distinct() %>% 
  glimpse()

```


Loading Analysis Results

```{r}
# clean  up format of MAST results
format_mastdf <- function(mast_df, name_col, level_col) {
  mast_df %>% 
    dplyr::rename(cell_type_name = {{name_col}}) %>% 
    janitor::clean_names() %>% 
    mutate(cell_type_level = {{level_col}}) %>% 
    glimpse()
}

# MAST Results (signficant hits only)
class_mast_df <- readRDS(glue("{data_path}/mast/snRNASeq/class_mast_signif_hits_2026-01-04.rds")) %>% 
    format_mastdf(name_col = "class_name", level_col = "class")
subclass_mast_df <- readRDS(
  glue("{data_path}/mast/snRNASeq/subclass_mast_signif_hits_2026-01-04.rds")) %>% 
    format_mastdf(name_col = "subclass_name", level_col = "subclass")
supertype_mast_df <- readRDS(
  glue("{data_path}/mast/snRNASeq/supertype_mast_signif_hits_2026-01-04.rds")) %>% 
    format_mastdf(name_col = "supertype_name", level_col = "supertype")
mast_df <- bind_rows(class_mast_df, subclass_mast_df, supertype_mast_df) %>% 
  filter(p_adjust < 0.1) %>% 
    glimpse()

# Augur Results
format_augurdf <- function(augur_df, name_col, level_col) {
  augur_df %>% 
    dplyr::rename(cell_type_name = {{name_col}}) %>% 
    mutate(cell_type_level = {{level_col}}) %>% 
    glimpse()
}

augur_resdir <- glue("{data_path}/augur/snRNASeq/var_thresh_0.6")
class_aucs_df <- readRDS(
  glue("{augur_resdir}/class_aucs_df_2026-01-10.rds")) %>% 
  format_augurdf(name_col = "class_name", level_col = "class")
subclass_aucs_df <- readRDS(
  glue("{data_path}/augur/snRNASeq/var_thresh_0.6/subclass_aucs_df_2026-01-10.rds")) %>% 
  format_augurdf(name_col = "subclass_name", level_col = "subclass")
supertype_aucs_df <- readRDS(
  glue("{data_path}/augur/snRNASeq/var_thresh_0.6/supertype_aucs_df_2026-01-10.rds")) %>% 
  format_augurdf(name_col = "supertype_name", level_col = "supertype")
augur_df <- bind_rows(class_aucs_df, subclass_aucs_df, supertype_aucs_df) %>% 
    glimpse()

# PseudoBulk Results (EdgeR LRT)
edgeR_res <- readRDS(
  glue("{data_path}/edgeR/snRNASeq/edgeR_signif_hits_lrt_res_2026-01-03.rds")) %>% 
  mutate(
    tissue = str_after_last(celltype_tissue, "_"), 
    cell_type = str_before_last(celltype_tissue, "_"), 
    cell_type_label = cell_type) %>% 
  filter(FDR < 0.1) %>% 
  # filter(PValue < 0.05) %>% 
  left_join(allen_annot_long) %>% 
  glimpse()

# Deseq2 Results
deseq2_res <- readRDS(
  glue("{data_path}/DESeq2/snRNASeq/deseq2_signif_hits_res_2026-01-03.rds")) %>% 
  mutate(
    tissue = str_after_last(celltype_tissue, "_"), 
    cell_type = str_before_last(celltype_tissue, "_"), 
    cell_type_label = cell_type) %>% 
  left_join(allen_annot_long) %>% 
  filter(padj < 0.1) %>% 
  left_join(allen_annot_long) %>% 
  glimpse()

```


Calculating the Odds ratio of each cell type for WildR vs SPF

```{r}

# (a/c) / (b/d)
# where a is the number of aged cells in the cluster, b is the number of adult cells in the cluster, c is the number of aged cells in the class minus a and d is the number of adult cells in the class minus b
# cell_type_name <- "class_name"

calculate_odds_ratio <- function(cell_type_name) {
  cell_counts_df <- seurat_obj@meta.data %>% 
    dplyr::select(cell_type_name = {{cell_type_name}}, microbiome, tissue) %>% 
    group_by(cell_type_name, microbiome, tissue) %>% 
    summarise(n_cells = n(), .groups = "drop")
    
  tissue_group_totals <- cell_counts_df %>% 
    group_by(tissue, microbiome) %>% 
    summarise(group_tissue_ncells = sum(n_cells), .groups = "drop") %>% 
    pivot_wider(names_from = microbiome, values_from = group_tissue_ncells, 
      names_prefix = "total_cells_") %>% 
    glimpse()

  cell_counts_df %>% 
    pivot_wider(names_from = microbiome, values_from = n_cells) %>% 
    left_join(tissue_group_totals) %>% 
    mutate(celltype_tissue = glue("{cell_type_name}_{tissue}")) %>% 
    mutate(odds_ratio = (WildR / (total_cells_WildR - WildR)) / (SPF / (total_cells_SPF - SPF))) %>% 
    glimpse()
}

# tst <- calculate_odds_ratio("class_name") %>% glimpse()
odds_ratio_df <- bind_rows(
  calculate_odds_ratio("class_name") %>% mutate(cell_type_level = "class"),
  calculate_odds_ratio("subclass_name") %>% mutate(cell_type_level = "subclass"),
  calculate_odds_ratio("supertype_name") %>% mutate(cell_type_level = "supertype")
) %>% 
arrange(desc(odds_ratio))
# View(odds_ratio_df)




```

## Q1: What cell types have the most DEGs, and are the most altered in state between WildR vs SPF?


```{r}
mast_df %>% glimpse()
edgeR_res %>% glimpse()
deseq2_res %>% glimpse()
augur_df %>% glimpse()

mast_deg_summary <- mast_df %>% 
    group_by(cell_type_name, tissue, cell_type_level) %>% 
    tally() %>% 
    mutate(analysis = "MAST")

edgeR_deg_summary <- edgeR_res %>% 
    group_by(cell_type_name, tissue, cell_type_level, celltype_tissue) %>% 
    tally() %>% 
    mutate(analysis = "EdgeR-Pseudobulk")

deseq2_deg_summary <- deseq2_res %>% 
    group_by(cell_type_name, tissue, cell_type_level, celltype_tissue) %>% 
    tally() %>% 
    mutate(analysis = "DESeq2-Pseudobulk")

```


```{r}
# Visualize a ranked list of cell types by number of DEGS in each tissue for each analysis
tiss <- "AMY"

deg_summary_plot <- bind_rows(mast_deg_summary, edgeR_deg_summary, deseq2_deg_summary) %>% 
  filter(tissue == tiss) %>% 
  filter(n > 1) %>% 
  mutate(analysis = factor(analysis, 
    levels = c("MAST", "EdgeR-Pseudobulk", "DESeq2-Pseudobulk"))) %>% 
  group_by(cell_type_level) %>% 
  arrange(n) %>% 
  mutate(cell_type_name = fct_reorder(cell_type_name, n)) %>% 
  glimpse()

# TEMP DELETE ME LATER
# deg_summary_plot %>% 
#   filter(analysis == "EdgeR-Pseudobulk") %>% 
#   View()

# Get intersection of cell types present in both datasets
augur_tiss_celltypes <- augur_df %>% 
  filter(tissue == tiss) %>% 
  pull(cell_type_name) %>% 
  unique()

# Get the ordered cell types from DEG plot, preserving order within each level
deg_celltypes_ordered <- deg_summary_plot %>% 
  ungroup() %>% 
  filter(analysis == "MAST") %>% 
  select(cell_type_name, cell_type_level) %>% 
  distinct() %>% 
  mutate(cell_type_name = as.character(cell_type_name))

# Filter to only common cell types, preserving the order
common_celltypes_df <- deg_celltypes_ordered %>% 
  filter(cell_type_name %in% augur_tiss_celltypes)

# Create DEG summary plot (filtered to common cell types, preserving order within each level)
p_deg_summary <- deg_summary_plot %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ungroup() %>%
  ggplot(aes(x = n, y = cell_type_name)) +
    geom_bar(stat = "identity", fill = "gray50", width = 0.75) +
    # scale_x_log10() +
    theme_bw() +
    facet_grid(cell_type_level ~ analysis, scales = "free", space = "free_y") +
    labs(x = "Number of DEGs (P-adjusted < 0.1)", y = "Cell Type") +
    theme(strip.text.y = element_blank(), 
      axis.text.x = element_text(angle = 45, hjust = 1))

# Create Augur lollipop chart (preserving order within each level)
p_augur_scores <- augur_df %>% 
  filter(tissue == tiss) %>% 
  mutate(analysis = "Augur") %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ggplot(aes(x = auc, y = cell_type_name)) +
    geom_segment(aes(x = 0.5, xend = auc, y = cell_type_name, yend = cell_type_name), 
                 color = "gray50", linewidth = 0.5) +
    geom_point(aes(fill = auc), size = 3, shape = 21) +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0.5) +
    scale_x_continuous(limits = c(0.5, 0.8), expand = c(0, 0)) +
    theme_bw() +
    facet_grid(cell_type_level ~  analysis, scales = "free", space = "free_y") +
    labs(x = "AUROC", y = "", color = "AUROC") +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          legend.position = "none"
          )

# Align plots using patchwork (which should align by shared y-axis)
p_combined <- p_deg_summary + p_augur_scores + 
  plot_layout(widths = c(3, 1), guides = "collect") +
  plot_annotation(title = "Differential Expression Analysis",
                  subtitle = tiss)
# p_combined

ggsave(
  glue("{fig_path}/DEG/snRNASeq/mast_deg_summary_combined_{tiss}_{Sys.Date()}.png"), 
  p_combined, width = 12, height = 16, dpi = 300)

```

Summary for HYPOTHALAMUS

```{r}
tiss <- "HYP"

deg_summary_plot <- bind_rows(mast_deg_summary, edgeR_deg_summary, deseq2_deg_summary) %>% 
  filter(tissue == tiss) %>% 
  filter(n > 1) %>% 
  mutate(analysis = factor(analysis, 
    levels = c("MAST", "EdgeR-Pseudobulk", "DESeq2-Pseudobulk"))) %>% 
  group_by(cell_type_level) %>% 
  arrange(n) %>% 
  left_join(allen_annot_long) %>% 
  # mutate(cell_type_name = fct_reoder(cell_type_name, cell_type_name)) %>% 
  glimpse()

# Get intersection of cell types present in both datasets
augur_tiss_celltypes <- augur_df %>% 
  filter(tissue == tiss) %>% 
  pull(cell_type_name) %>% 
  unique()

# Get the ordered cell types from DEG plot, preserving order within each level
deg_celltypes_ordered <- deg_summary_plot %>% 
  ungroup() %>% 
  filter(analysis == "MAST") %>% 
  select(cell_type_name, cell_type_level) %>% 
  distinct() %>% 
  mutate(cell_type_name = as.character(cell_type_name))

# Filter to only common cell types, preserving the order
common_celltypes_df <- deg_celltypes_ordered %>% 
  filter(cell_type_name %in% augur_tiss_celltypes)

# Create DEG summary plot (filtered to common cell types, preserving order within each level)
p_deg_summary <- deg_summary_plot %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ungroup() %>%
  ggplot(aes(x = n, y = cell_type_name)) +
    geom_bar(stat = "identity", fill = "gray50", width = 0.75) +
    # scale_x_log10() +
    theme_bw() +
    facet_grid(cell_type_level ~ analysis, scales = "free", space = "free_y") +
    labs(x = "Number of DEGs (P-adjusted < 0.1)", y = "Cell Type") +
    theme(strip.text.y = element_blank(), 
      axis.text.x = element_text(angle = 45, hjust = 1))

# Create Augur lollipop chart (preserving order within each level)
p_augur_scores <- augur_df %>% 
  filter(tissue == tiss) %>% 
  mutate(analysis = "Augur") %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ggplot(aes(x = auc, y = cell_type_name)) +
    geom_segment(aes(x = 0.5, xend = auc, y = cell_type_name, yend = cell_type_name), 
                 color = "gray50", linewidth = 0.5) +
    geom_point(aes(fill = auc), size = 3, shape = 21) +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0.5) +
    scale_x_continuous(limits = c(0.5, 0.8), expand = c(0, 0)) +
    theme_bw() +
    facet_grid(cell_type_level ~  analysis, scales = "free_y", space = "free") +
    labs(x = "AUROC", y = "", color = "AUROC") +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          legend.position = "none"
          )

# Align plots using patchwork (which should align by shared y-axis)
p_combined <- p_deg_summary + p_augur_scores + 
  plot_layout(widths = c(3, 1), guides = "collect") +
  plot_annotation(title = "Differential Expression Analysis",
                  subtitle = tiss)

# p_combined

ggsave(
  glue("{fig_path}/DEG/snRNASeq/mast_deg_summary_combined_{tiss}_{Sys.Date()}.png"), 
  p_combined, width = 12, height = 19, dpi = 300)


```



REMAKING SUMMARY FIGURES WITH BOTH Pvalue and FDR thresholds

```{r}
mast_df_pvalue <- bind_rows(class_mast_df, subclass_mast_df, supertype_mast_df) %>% 
  filter(pr_chisq < 0.05) %>% 
    glimpse()

edgeR_res_pvalue <- readRDS(
  glue("{data_path}/edgeR/snRNASeq/edgeR_signif_hits_lrt_res_2026-01-03.rds")) %>% 
  mutate(
    tissue = str_after_last(celltype_tissue, "_"), 
    cell_type = str_before_last(celltype_tissue, "_"), 
    cell_type_label = cell_type) %>% 
  filter(PValue < 0.05) %>% 
  left_join(allen_annot_long) %>% 
  glimpse()

# Deseq2 Results
deseq2_res_pvalue <- readRDS(
  glue("{data_path}/DESeq2/snRNASeq/deseq2_signif_hits_res_2026-01-03.rds")) %>% 
  mutate(
    tissue = str_after_last(celltype_tissue, "_"), 
    cell_type = str_before_last(celltype_tissue, "_"), 
    cell_type_label = cell_type) %>% 
  left_join(allen_annot_long) %>% 
  filter(pvalue < 0.05) %>% 
  left_join(allen_annot_long) %>% 
  glimpse()

mast_deg_summary_pvalue <- mast_df_pvalue %>% 
    group_by(cell_type_name, tissue, cell_type_level) %>% 
    tally() %>% 
    mutate(analysis = "MAST")
edgeR_deg_summary_pvalue <- edgeR_res_pvalue %>% 
    group_by(cell_type_name, tissue, cell_type_level, celltype_tissue) %>% 
    tally() %>% 
    mutate(analysis = "EdgeR-Pseudobulk")
deseq2_deg_summary_pvalue <- deseq2_res_pvalue %>% 
    group_by(cell_type_name, tissue, cell_type_level, celltype_tissue) %>% 
    tally() %>% 
    mutate(analysis = "DESeq2-Pseudobulk")

```

  

```{r}
# Visualize a ranked list of cell types by number of DEGS in each tissue for each analysis
tiss <- "AMY"

deg_summary_plot <- bind_rows(mast_deg_summary, edgeR_deg_summary, deseq2_deg_summary) %>% 
  filter(tissue == tiss) %>% 
  filter(n > 1) %>% 
  mutate(analysis = factor(analysis, 
    levels = c("MAST", "EdgeR-Pseudobulk", "DESeq2-Pseudobulk"))) %>% 
  group_by(cell_type_level) %>% 
  arrange(n) %>% 
  mutate(pthres = "FDR < 0.1") %>% 
  glimpse()

deg_summary_plot_pval <- bind_rows(
    mast_deg_summary_pvalue, edgeR_deg_summary_pvalue, deseq2_deg_summary_pvalue) %>% 
    left_join(allen_annot_long) %>% 
  filter(tissue == tiss) %>% 
  filter(cell_type_name %in% deg_summary_plot$cell_type_name) %>% 
  mutate(analysis = factor(analysis, 
    levels = c("MAST", "EdgeR-Pseudobulk", "DESeq2-Pseudobulk"))) %>% 
  mutate(pthres = "P-value < 0.05") %>% 
  glimpse()

# Get intersection of cell types present in both datasets
augur_tiss_celltypes <- augur_df %>% 
  filter(tissue == tiss) %>% 
  pull(cell_type_name) %>% 
  unique()

# Get the ordered cell types from DEG plot, preserving order within each level
deg_celltypes_ordered <- deg_summary_plot %>% 
  ungroup() %>% 
  filter(analysis == "MAST") %>% 
  select(cell_type_name, cell_type_level) %>% 
  distinct() %>% 
  mutate(cell_type_name = as.character(cell_type_name))

# Filter to only common cell types, preserving the order
common_celltypes_df <- deg_celltypes_ordered %>% 
  filter(cell_type_name %in% augur_tiss_celltypes)

# Create DEG summary plot (filtered to common cell types, preserving order within each level)
p_deg_summary <- bind_rows(deg_summary_plot_pval, deg_summary_plot) %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  # mutate(pthres = factor(pthres, levels = c("FDR < 0.1", "P-value < 0.05"))) %>% 
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ungroup() %>%
  ggplot(aes(x = n, y = cell_type_name, fill = pthres)) +
    geom_bar(stat = "identity", position = "identity", width = 0.8) +
    scale_fill_manual(values = c("P-value < 0.05" = "darkgrey", "FDR < 0.1" = "darkred")) +
    scale_x_log10() +
    theme_bw() +
    facet_grid(cell_type_level ~ analysis, scales = "free", space = "free_y") +
    labs(x = "Number of DEGs", y = "Cell Type", fill = NULL) +
    theme(strip.text.y = element_blank(), 
      axis.text.x = element_text(angle = 45, hjust = 1))

# Create Augur lollipop chart (preserving order within each level)
p_augur_scores <- augur_df %>% 
  filter(tissue == tiss) %>% 
  mutate(analysis = "Augur") %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ggplot(aes(x = auc, y = cell_type_name)) +
    geom_segment(aes(x = 0.5, xend = auc, y = cell_type_name, yend = cell_type_name), 
                 color = "gray50", linewidth = 0.5) +
    geom_point(aes(fill = auc), size = 3, shape = 21) +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0.5) +
    scale_x_continuous(limits = c(0.5, 0.8), expand = c(0, 0)) +
    theme_bw() +
    facet_grid(cell_type_level ~  analysis, scales = "free", space = "free_y") +
    labs(x = "AUROC", y = "", color = "AUROC") +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          legend.position = "none"
          )

# Align plots using patchwork (which should align by shared y-axis)
p_combined <- p_deg_summary + p_augur_scores + 
  plot_layout(widths = c(3, 1), guides = "collect") +
  plot_annotation(title = "Differential Expression Analysis",
                  subtitle = tiss)
# p_combined

ggsave(
  glue("{fig_path}/DEG/snRNASeq/mast_deg_summary_combined_{tiss}_pvalue_fdr_{Sys.Date()}.png"), 
  p_combined, width = 12, height = 16, dpi = 300)

```

Summary for HYPOTHALAMUS

```{r}
tiss <- "HYP"

deg_summary_plot <- bind_rows(mast_deg_summary, edgeR_deg_summary, deseq2_deg_summary) %>% 
  filter(tissue == tiss) %>% 
  filter(n > 1) %>% 
  mutate(analysis = factor(analysis, 
    levels = c("MAST", "EdgeR-Pseudobulk", "DESeq2-Pseudobulk"))) %>% 
  group_by(cell_type_level) %>% 
  arrange(n) %>% 
  mutate(pthres = "FDR < 0.1") %>% 
  glimpse()

deg_summary_plot_pval <- bind_rows(
    mast_deg_summary_pvalue, edgeR_deg_summary_pvalue, deseq2_deg_summary_pvalue) %>% 
    left_join(allen_annot_long) %>% 
  filter(tissue == tiss) %>% 
  filter(cell_type_name %in% deg_summary_plot$cell_type_name) %>% 
  mutate(analysis = factor(analysis, 
    levels = c("MAST", "EdgeR-Pseudobulk", "DESeq2-Pseudobulk"))) %>% 
  mutate(pthres = "P-value < 0.05") %>% 
  glimpse()

# Get intersection of cell types present in both datasets
augur_tiss_celltypes <- augur_df %>% 
  filter(tissue == tiss) %>% 
  pull(cell_type_name) %>% 
  unique()

# Get the ordered cell types from DEG plot, preserving order within each level
deg_celltypes_ordered <- deg_summary_plot %>% 
  ungroup() %>% 
  filter(analysis == "MAST") %>% 
  select(cell_type_name, cell_type_level) %>% 
  distinct() %>% 
  mutate(cell_type_name = as.character(cell_type_name))

# Filter to only common cell types, preserving the order
common_celltypes_df <- deg_celltypes_ordered %>% 
  filter(cell_type_name %in% augur_tiss_celltypes)

# Create DEG summary plot (filtered to common cell types, preserving order within each level)
p_deg_summary <- bind_rows(deg_summary_plot_pval, deg_summary_plot) %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  # mutate(pthres = factor(pthres, levels = c("FDR < 0.1", "P-value < 0.05"))) %>% 
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ungroup() %>%
  ggplot(aes(x = n, y = cell_type_name, fill = pthres)) +
    geom_bar(stat = "identity", position = "identity", width = 0.7) +
    scale_fill_manual(values = c("P-value < 0.05" = "darkgrey", "FDR < 0.1" = "darkred")) +
    scale_x_log10() +
    theme_bw() +
    facet_grid(cell_type_level ~ analysis, scales = "free", space = "free_y") +
    labs(x = "Number of DEGs", y = "Cell Type", fill = NULL) +
    theme(strip.text.y = element_blank(), 
      axis.text.x = element_text(angle = 45, hjust = 1))

# Create Augur lollipop chart (preserving order within each level)
p_augur_scores <- augur_df %>% 
  filter(tissue == tiss) %>% 
  mutate(analysis = "Augur") %>% 
  filter(cell_type_name %in% common_celltypes_df$cell_type_name) %>%
  mutate(cell_type_name = factor(cell_type_name, 
                                 levels = deg_celltypes_ordered$cell_type_name)) %>%
  ggplot(aes(x = auc, y = cell_type_name)) +
    geom_segment(aes(x = 0.5, xend = auc, y = cell_type_name, yend = cell_type_name), 
                 color = "gray50", linewidth = 0.5) +
    geom_point(aes(fill = auc), size = 3, shape = 21) +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0.5) +
    scale_x_continuous(limits = c(0.5, 0.8), expand = c(0, 0)) +
    theme_bw() +
    facet_grid(cell_type_level ~  analysis, scales = "free", space = "free_y") +
    labs(x = "AUROC", y = "", color = "AUROC") +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          legend.position = "none"
          )

# Align plots using patchwork (which should align by shared y-axis)
p_combined <- p_deg_summary + p_augur_scores + 
  plot_layout(widths = c(3, 1), guides = "collect") +
  plot_annotation(title = "Differential Expression Analysis",
                  subtitle = tiss)
# p_combined

ggsave(
  glue("{fig_path}/DEG/snRNASeq/mast_deg_summary_combined_{tiss}_pvalue_fdr_{Sys.Date()}.png"), 
  p_combined, width = 12, height = 19, dpi = 300)


```



Spot checking that tissue labels are correct for the EdgeR data - it somehow seems that EdgeR trends in HYP match whats going on in MAST(AMY) data better than in MAST (HYP) data


```{r}
edger_dir <- glue("{data_path}/edgeR/snRNASeq")
mast_resdir <- glue("{data_path}/mast/snRNASeq")
# View(allen_annot_df)
# CS20230722_CLAS_12_AMY

mast_12_hyp <- readRDS(glue("{mast_resdir}/class_label/CS20230722_CLAS_12_HYP_2025-12-31.rds")) %>% 
  dplyr::select(gene_id = primerid, PValue = `Pr(>Chisq)`, FDR = p_adjust, coef) %>% 
    dplyr::rename_with(~ paste0(.x, "_HYP_MAST"), .cols = c(PValue, FDR, coef)) %>% 
  glimpse()

pb_12_hyp <- readRDS(glue("{edger_dir}/class_label/CS20230722_CLAS_12_HYP/res_lrt_2025-12-30.rds")) %>% 
  select(logFC, PValue, FDR) %>% 
  dplyr::rename_all(~ paste0(.x, "_HYP")) %>% 
  rownames_to_column("gene_id") %>% 
  glimpse()
pb_12_amy <- readRDS(glue("{edger_dir}/class_label/CS20230722_CLAS_12_AMY/res_lrt_2025-12-30.rds")) %>% 
  select(logFC, PValue, FDR) %>% 
  dplyr::rename_all(~ paste0(.x, "_AMY")) %>% 
  rownames_to_column("gene_id") %>% 
  glimpse()


merged_12_data <- full_join(pb_12_hyp, pb_12_amy, by = "gene_id") %>% 
  full_join(mast_12_hyp) %>% 
  glimpse()

# I want to know if the labels on the edgeR results are correct for tissue, looking to see if the results with MAST data are more consistent in the right tissue (HYP) vs AMY
merged_12_data %>% 
  ggplot(aes(x = logFC_HYP, y = coef_HYP_MAST)) +
    geom_point() +
    theme_bw() +
    labs(x = "EdgeR LogFC", y = "MAST LogFC")
  
merged_12_data %>% 
  ggplot(aes(x = logFC_AMY, y = coef_HYP_MAST)) +
    geom_point() +
    theme_bw() +
    labs(x = "EdgeR LogFC", y = "MAST LogFC")


```



Plotting counts of significant hits across tissue groups

```{r}

mast_deg_summary_wide <- mast_deg_summary %>% dplyr::rename(n_mast = n) %>% select(-analysis) 
edgeR_deg_summary_wide <- edgeR_deg_summary %>% dplyr::rename(n_edgeR = n) %>% select(-analysis) 

mast_edger_summary <- full_join(
  mast_deg_summary_wide, 
  edgeR_deg_summary_wide) %>% 
  glimpse()

mast_edger_summary %>% 
  ggplot(aes(x = n_mast, y = n_edgeR)) +
    geom_point() +
    theme_bw() +
    scale_x_log10() +
    scale_y_log10() +
    geom_smooth(method = "lm") +
    facet_wrap(~ tissue) +
    labs(x = "Number of DEGs (MAST)", y = "Number of DEGs (EdgeR)")

# swapping tissue labels to test if its a better fit 
edgeR_deg_summary_wide_swapped <- edgeR_deg_summary %>% dplyr::rename(n_edgeR = n) %>% select(-analysis)  %>% 
  mutate(tissue = case_when(
    tissue == "AMY" ~ "HYP",
    tissue == "HYP" ~ "AMY",
    TRUE ~ tissue
  ))

mast_edger_summary_swapped <- full_join(
  mast_deg_summary_wide, 
  edgeR_deg_summary_wide_swapped) %>% 
  glimpse()

mast_edger_summary_swapped %>% 
  ggplot(aes(x = n_mast, y = n_edgeR)) +
    geom_point() +
    theme_bw() +
    scale_x_log10() +
    scale_y_log10() +
    geom_smooth(method = "lm") +
    facet_wrap(~ tissue) +
    labs(x = "Number of DEGs (MAST)", y = "Number of DEGs (EdgeR)")



```



### Q1.A Is there any trend between how many cells are used in MAST and the number of DEGs? 

Some cell types appear to have a bunch of differences, is this just because of greater cell counts going into the analysis?

```{r}
library(ggpubr)
library(ggrepel)

get_mast_cell_counts <- function(cell_type_column, celltype, tissue)  {
    get_time <- function() print(format(Sys.time(), "%Y-%m-%d_%H:%M:%S"))
    # Filter CELLS using metadata directly
    message(glue("{get_time()} Filtering cells"))
    cells_to_keep <- seurat_obj@meta.data[[cell_type_column]] == celltype & 
                    seurat_obj@meta.data$tissue == tissue
    return(sum(cells_to_keep))
}

celltype_droplist <- readRDS(
  glue("{data_path}/seurat/snRNASeq/celltype_droplist_2025-12-30.rds")
  )

metadata_df <- seurat_obj@meta.data %>% glimpse()
class_batchdata <- 
    dplyr::select(metadata_df, celltype = class_label, tissue) %>% 
    distinct() %>%
    mutate(cell_type_column = "class_label")
subclass_batchdata <- 
    dplyr::select(metadata_df, celltype = subclass_label, tissue) %>% 
    distinct() %>%
    mutate(cell_type_column = "subclass_label")
supertype_batchdata <- 
    dplyr::select(metadata_df, celltype = supertype_label, tissue) %>% 
    distinct() %>%
    mutate(cell_type_column = "supertype_label")

batchparams_df <- 
    bind_rows(class_batchdata, subclass_batchdata, supertype_batchdata) %>% 
    left_join(celltype_droplist) %>%
    filter(is.na(dropme)) %>% 
    dplyr::select(cell_type_column, celltype, tissue) %>% 
    glimpse()

mast_cell_counts_df <- batchparams_df %>%
  mutate(n_cells = pmap_dbl(list(cell_type_column, celltype, tissue), get_mast_cell_counts)) %>%
  mutate(cell_type_label = celltype) %>% 
  left_join(allen_annot_long) %>% 
  glimpse()

# new update subsamples down to 2000 cells per group max
mast_cell_counts_df %<>%
  mutate(n_cells = if_else(n_cells > 4000, 4000, n_cells)) %>% 
  glimpse()

mast_deg_summary_strict <- bind_rows(class_mast_df, subclass_mast_df, supertype_mast_df) %>%
    # filter(pr_chisq < 0.05) %>% 
    filter(p_adjust < 0.01) %>% 
    group_by(cell_type_name, tissue, cell_type_level, celltype_tissue) %>% 
    tally() %>% 
    mutate(analysis = "MAST") %>% 
    glimpse()

mast_deg_summary_wcounts <- mast_deg_summary_strict %>% 
  left_join(mast_cell_counts_df) %>% 
  glimpse()

p_mast_deg_summary_wcounts <- mast_deg_summary_wcounts %>% 
  ggplot(aes(x = n_cells, y = n)) +
    geom_point() +
    geom_smooth(method = "lm") +
    stat_cor(method = "pearson", label.x.npc = 0.05, label.y.npc = 0.97) +
    stat_regline_equation(label.x.npc = 0.05, label.y.npc = 0.91) +
    theme_bw() +
    # geom_vline(xintercept = 500, linetype = "dashed") +
    # scale_x_log10() +
    # scale_y_log10() +
    # facet_grid(cell_type_level ~ tissue) +
    facet_wrap( ~ tissue , scales = "free") +
    labs(x = "Number of Cells", y = "Number of DEGs")

ggsave(
  glue("{fig_path}/DEG/snRNASeq/mast_deg_summary_wcounts_{Sys.Date()}.png"), 
  p_mast_deg_summary_wcounts, width = 10, height = 6, dpi = 300)


# Identify top 5 cell types with most DEGs per tissue
top5_celltypes <- mast_deg_summary_wcounts %>%
  group_by(tissue) %>%
  slice_max(n, n = 10) %>%
  ungroup()

p_mast_deg_summary_wcounts_labeled <- mast_deg_summary_wcounts %>% 
  ggplot(aes(x = n_cells, y = n)) +
    geom_point() +
    geom_smooth(method = "lm") +
    stat_cor(method = "pearson", label.x.npc = 0.05, label.y.npc = 0.97) +
    stat_regline_equation(label.x.npc = 0.05, label.y.npc = 0.91) +
    geom_label_repel(
      data = top5_celltypes,
      aes(label = cell_type_name),
      alpha = 0.8,
      force = 10,
      force_pull = 0,
      nudge_x = 0.25,
      nudge_y = 0.25,
      segment.curvature = 0.3,
      segment.ncp = 3,
      segment.angle = 10,
      min.segment.length = 0,
      box.padding = 0.5,
      point.padding = 0.3,
      max.overlaps = Inf
    ) +
    theme_bw() +
    facet_wrap( ~ tissue , scales = "free") +
    labs(x = "Number of Cells", y = "Number of DEGs")

ggsave(
  glue("{fig_path}/DEG/snRNASeq/mast_deg_summary_wcounts_labeled_{Sys.Date()}.png"), 
  p_mast_deg_summary_wcounts_labeled, width = 10, height = 6, dpi = 300)

# s
# looks like theres a correlation in the HYP but not the AMY for number of cells in analysis and number of DEGs returned for that cell type with MAST.

```





```{r}

```

