---
title: "Mining Border-Associated Macrophages from Microglia-Enriched Mouse Brain Tissues Using Single-Cell RNA-Seq"
editor: visual
author: "Joe Boktor"
date: '2024-04-13'
format: 
  html:
    font-family: helvetica neue
    page-layout: full
    toc: true
    toc-location: left
    toc-depth: 5
    self-contained: true
    code-fold: false
    code-tools: true
    fig-align: center
    grid:
      sidebar-width: 200px
      body-width: 1100px
      margin-width: 200px
      gutter-width: 1.5em
---

##  Overview

In this notebook, we analyze single cell RNA-Seq data from [Abdel-Haq et al., 2022.](https://elifesciences.org/articles/81453) to identify Border Associated Macrophages (BAMs) in the Substantia Nigra and Striatum of WT and ASO Thy1-BDF1 mice, fed either a control or prebiotic rich diet. This analysis quantifies the number predicted BAMs in each tissue/condition, and further assess this cell population for alterations in gene expression and GO pathways.

##  Analysis Setup

Setting up environment.

```{r}
#| warning: false
#| message: false

library(tidyverse)
library(magrittr)
library(patchwork)
# library(spdep)
# library(Voyager)
library(SingleCellExperiment)
# library(SpatialExperiment)
# library(SpatialFeatureExperiment)
# library(batchelor)
library(scater)
# library(scran)
# library(bluster)
# library(fossil)
library(glue)
library(Seurat)
# library(SeuratDisk)
# library(SeuratData)
# library(scCustomize)
# library(glmGamPoi)
# library(spatialLIBD)
# library(reticulate)

# parallelization
library(BiocParallel)
library(future)
library(furrr)
# plotting
library(RColorBrewer)
library(ggsci)
library(Matrix)
library(plotly)
library(DT)
library(dplyr)
library(purrr)

# setting paths
homedir <- "/central/groups/mthomson/jboktor"
wkdir <- glue("{homedir}/spatial_genomics/jess_2024-01-23")
h5ad_dir <- glue("{wkdir}/data/input/temp")

```


Reading in scRNA-Seq data.
```{r, eval=FALSE}

sample_map <- c(
  # substantia nigra
  "Sample 1" = "Control WT",
  "Sample 2" = "Control ASO",
  "Sample 3" = "Prebiotic WT",
  "Sample 4" = "Prebiotic ASO",
  # striatum
  "Sample 5" = "Control WT",
  "Sample 6" = "Control ASO",
  "Sample 7" = "Prebiotic WT",
  "Sample 8" = "Prebiotic ASO"
)

## Converting h5ad to h5seurat | need to do this only once
Convert(glue("{h5ad_dir}/Striatum.h5ad"),
    dest = "h5seurat",
    overwrite = TRUE
)
Convert(glue("{h5ad_dir}/Substantia.h5ad"),
    dest = "h5seurat",
    overwrite = TRUE
)

# Reading in converted seurat objects
striatum <- LoadH5Seurat(glue("{h5ad_dir}/Striatum.h5seurat"))
striatum
nigra <- LoadH5Seurat(glue("{h5ad_dir}/Substantia.h5seurat"))
nigra
```


Here we follow the QC protocol in [Abdel-Haq et al., 2022.](https://elifesciences.org/articles/81453), where we first filter cells with fewer than 200 genes detected and genes which are expressed in fewer than 100 cells. Data are normalized and variance stabilized by SCTransform function from Seurat.


Function to format and perform QC filtering on raw cell x gene matrices.

```{r, eval=FALSE}
format_seurat_obj <- function(seurat_obj, sample_map) {
  # adding group to metadata
  seurat_obj$condition <- seurat_obj@meta.data$sample_id %>% purrr::map_chr(
    ~ sample_map[.x]
  )
  seurat_obj$group <- strex::str_after_first(
    seurat_obj$condition, " "
  )

  # create count and feature columns in metadata
  # create count and feature columns in metadata
  seurat_obj$nCount_RNA <-
    colSums(GetAssayData(object = seurat_obj, slot = "counts"))
  seurat_obj$nFeature_RNA <-
    colSums(GetAssayData(object = seurat_obj, slot = "counts") > 0)

  # Filter cells with fewer than 200 genes
  seurat_obj <- subset(seurat_obj, subset = nFeature_RNA > 200)
  # Filter genes that are expressed in fewer than 100 cells
  seurat_obj <- subset(seurat_obj,
    features = rowSums(seurat_obj@assays$RNA@counts > 0) > 100
  )
  seurat_obj <- seurat_obj %>%
    SCTransform(assay = "RNA") %>%
    RunPCA(npcs = 4, features = rownames(seurat_obj)) %>%
    RunUMAP(dims = 1:4) #%>%
    # FindNeighbors(reduction = "pca", dims = 1:4) %>%
    # FindClusters(resolution = 0.3)
  
  return(seurat_obj)
}
```

Applying the QC filtering function to the two datasets.

```{r, eval=FALSE}
nigra_clust <- format_seurat_obj(nigra, sample_map)
nigra_clust@meta.data %>% glimpse

striatum_clust <- format_seurat_obj(striatum, sample_map)
striatum_clust@meta.data %>% glimpse

saveRDS(
  nigra_clust,
  glue("{wkdir}/data/input/temp/seurat_substantia_nigra_qc.rds")
)
saveRDS(
  striatum_clust,
  glue("{wkdir}/data/input/temp/seurat_striatum_qc.rds")
)

```


## Quantifying BAMs 

```{r, eval=FALSE}
nigra_clust <- readRDS(
  glue("{wkdir}/data/input/temp/seurat_substantia_nigra_qc.rds")
)
striatum_clust <- readRDS(
  glue("{wkdir}/data/input/temp/seurat_striatum_qc.rds")
)

```

Plotting expression levels of key genes of interest for BAMs across UMAP profiles.

```{r, eval=FALSE}
# Plotting UMAPs
# checking matching name symbols
data.frame(
  "symbols" = rownames(striatum_clust)) %>%
  mutate(symbols_up = toupper(symbols)) %>%
    filter(symbols_up %in%
      c(
        "CX3CR1", "ClEC12A", "MS4A7", "H2-AA", "CD38",
        "PF4", "CD163", "MRC1", "GPNMB",
        "CXCL10", "Il1B", "CD274", "CD86", "CD74",
        "CD206", "LYVE1", "VEGFR1", "FLT1", 
        # Microglia marker
        "TMEM119"
      )) %>% 
      arrange(symbols)


p_striatum_umap <- FeaturePlot(striatum_clust,
  features = c("Cxcl10", "Cx3cr1", "H2-Aa", "Mrc1",
  "Ms4a7", "Gpnmb", "Lyve1",
  "Pf4", "Cd86", "Cd74", "Cd274", "Flt1"),
  split.by = "group",
  order = TRUE,
  pt.size	= 0.5,
  reduction = "umap"
)

ggsave(
  glue("{wkdir}/figures/temp/umap_gene_highlights_STRIATUM.png"),
  p_striatum_umap,
  width = 9,
  height = 30
)

p_nigra_umap <- FeaturePlot(nigra_clust,
  features = c("Cxcl10", "Cx3cr1", "H2-Aa", "Mrc1",
  "Ms4a7", "Gpnmb", "Lyve1",
  "Pf4", "Cd86", "Cd74", "Cd274", "Flt1"),
  split.by = "group",
  order = TRUE,
  pt.size	= 0.5,
  reduction = "umap"
)

ggsave(
  glue("{wkdir}/figures/temp/umap_gene_highlights_NIGRA.png"),
  p_nigra_umap,
  width = 9,
  height = 30
)

```


### Substantia Nigra Profiles 
![](../figures/temp/umap_gene_highlights_NIGRA.png)

### Striatum Profiles 
![](../figures/temp/umap_gene_highlights_STRIATUM.png)

Next, we filter out cells that are predicted to be BAMs based on any expression level of key BAM-marker genes (Mrc1, Flt1, Pf4, and Cd163), and low expression of microglia specific marker (Tmem119).


```{r, eval=FALSE}
#' filters cells with any expression level of core BAM marker genes 
filter_bams <- function(seurat_obj) {
  # Filter cells with at least 10 counts in any of the specified genes
  DefaultAssay(seurat_obj) <- "RNA"
  filtered_cells <- subset(seurat_obj,
    subset = Mrc1 > 0 | Flt1 > 0 | Pf4 > 0 | Cd163 > 0 | Gpnmb > 0 | Ms4a7 > 0 
  )
  filtered_cells <- subset(filtered_cells,
    subset = Tmem119 < 5
  )
  DefaultAssay(filtered_cells) <- "SCT"
  return(filtered_cells)
}

nigra_bams <- filter_bams(nigra_clust)
striatum_bams <- filter_bams(striatum_clust)

glue(
  "In the Striatum there are {nrow(striatum_bams@meta.data)} ",
  "possible BAMs out of {nrow(striatum_clust@meta.data)} cells"
) %>% message()

glue(
  "In the Substantia Nigra there are {nrow(nigra_bams@meta.data)} ",
  "possible BAMs out of {nrow(nigra_clust@meta.data)} cells"
) %>% message()

```



- In the Striatum there are **58** possible BAMs out of **27,152** cells.

- In the Substantia Nigra there are **16** possible BAMs out of **5,278** cells


Next, we summarize the fraction of predicted BAMs within group (ASO/WT) (Control/Prebiotic) in the SN and striatum

```{r, eval=FALSE}
#' Summarizes counts of cells that are possible BAMs by groups
summarize_bam_counts <- function(seurat_obj, seurat_obj_filtered) {
  original_metadata <- seurat_obj@meta.data
  filtered_metadata <- seurat_obj_filtered@meta.data

  original_counts <- original_metadata %>%
    dplyr::count(condition) %>%
    rename(total_cells = n)

  filtered_counts <- filtered_metadata %>%
    dplyr::count(condition) %>%
    rename(predicted_bams = n)

  counts <- original_counts %>%
    left_join(filtered_counts, by = "condition") %>%
    mutate(
      percent_BAMs = round((predicted_bams / total_cells) * 100, digits = 2)
    )
  return(counts)
}

bam_stats <- list()
bam_stats[["nigra_bam_stats"]] <- summarize_bam_counts(nigra_clust, nigra_bams)
bam_stats[["sn_bam_stats"]] <- summarize_bam_counts(striatum_clust, striatum_bams)

saveRDS(
  bam_stats,
  glue("{wkdir}/data/input/temp/bam_summary_stats.rds")
)

```


```{r}
bam_stats <- readRDS(
  glue("{wkdir}/data/input/temp/bam_summary_stats.rds")
)

bam_stats[["nigra_bam_stats"]] %>%
  datatable(options = list(pageLength = 5, autoWidth = TRUE))
bam_stats[["sn_bam_stats"]] %>%
  datatable(options = list(pageLength = 5, autoWidth = TRUE))

p_str_bam_barplot <- bam_stats[["sn_bam_stats"]] %>%
  ggplot(aes(x = condition, y = percent_BAMs)) +
  geom_col(width = 0.6) +
   labs(x = NULL, y = "Relative proportion of predicted BAMs (%)") +
   theme_bw() +
   theme(axis.text.x = element_text(angle=45, hjust=1))

ggsave(
  glue("{wkdir}/figures/temp/striatum_bam_relab.png"),
  p_str_bam_barplot, 
  width = 3, height = 5
)

```


![Relative Abundance of BAMs in the Striatum](../figures/temp/striatum_bam_relab.png){height=600px}


### Key Takeaways

- The Striatum displays a subtly but quantifiable signal for BAMs, likely related to the overal larger number of cells profiled compared to the SN. 
- In the Striatum, there is a 3-fold increase in the fraction of BAMs per cell isolate population in ASO mice compared to WT, showing a genotype but not diet effect.
- The Substantia Nigra shows limited and likely unreliable signals for BAMs.



## Differential gene expression analysis of BAMs across genotype and diet conditions.

Finding differentially expressed genes within BAMs across groups.

```{r, eval=FALSE}

Idents(striatum_bams) <- "condition"
group_comparisons <- as.data.frame(t(combn(unique(sample_map), 2)))

markers_df <- purrr::map2(
  group_comparisons$V1,
  group_comparisons$V2,
  ~ FindMarkers(striatum_bams,
    ident.1 = .x, ident.2 = .y,
    min.pct = 0.25, logfc.threshold = 0.25
  ) %>%
    mutate(Var1 = .x, Var2 = .y, comparison = glue("{Var1} - {Var2}")) %>%
    rownames_to_column(var="gene_symbol")
) %>%
  bind_rows() %>%
  glimpse()


markers_df %>%
  filter(p_val_adj <= 0.1) %>%
  group_by(comparison) %>%
  dplyr::count()

markers_df %>%
  filter(p_val <= 0.05) %>%
  group_by(comparison) %>%
  dplyr::count()

saveRDS(
  markers_df,
  glue("{wkdir}/data/input/temp/BAM_DEG_markers.rds")
)

```


When using only a p-value threshold of 0.05, we see the following number of genes of interest across groups.

| Comparison                       | n   |
|----------------------------------|-----|
| Control ASO - Prebiotic ASO      | 189 |
| Control ASO - Prebiotic WT       | 370 |
| Control WT - Control ASO         | 105 |
| Control WT - Prebiotic ASO       | 167 |
| Control WT - Prebiotic WT        | 139 |
| Prebiotic WT - Prebiotic ASO     | 458 |


However, if we adjust use adjusted p-values and set a threshold of 0.1, we only observe two comparisons with a significant difference and each only contains a single gene:


| Comparison                       | n  | Gene     |
|----------------------------------|----|----------|
| Control ASO - Prebiotic ASO      | 1  | Sfi1     |
| Prebiotic WT - Prebiotic ASO     | 1  | Fam160b1 |


### Gene-Ontology Analysis

Here we select genes with a non-adjusted pvalue of 0.05 and perform gene ontology analysis to identify enriched biological processes, cellular components, and molecular functions. 

```{r, eval=FALSE}
library(clusterProfiler)
library(org.Mm.eg.db)

markers_df <- readRDS(
  glue("{wkdir}/data/input/temp/BAM_DEG_markers.rds")
)

deg_df <- markers_df %>%
  filter(p_val <= 0.05) %>%
  mutate(direction = if_else(
    avg_log2FC > 0, "Upregulated", "Downregulated"
  ))

deg_df %>% glimpse


# format params for GO enrichment
params_1 <- expand.grid(
  unique(deg_df$comparison),
  c( "Upregulated", "Downregulated")
) %>% 
dplyr::rename(
  comparison = Var1,
  direction = Var2
)
params_2 <- expand.grid(
  c("BP", "CC", "MF"),
  c( "Upregulated", "Downregulated")
) %>% 
dplyr::rename(
  Ontology = Var1,
  direction = Var2
)
go_params <- params_1 %>% 
  full_join(params_2, by = "direction", relationship = "many-to-many") %>%
  glimpse()


future::plan("multisession", workers = 8)
enrichGO_res <-
  furrr::future_pmap(
    list(
      go_params$Ontology,
      go_params$direction,
      go_params$comparison
    ),
    ~ enrichGO(
      gene = deg_df %>%
        filter(direction == ..2 & comparison == ..3) %>%
        pull(gene_symbol),
      OrgDb = org.Mm.eg.db,
      keyType = "SYMBOL",
      ont = ..1,
      pAdjustMethod = "BH",
      pvalueCutoff = 0.05,
      qvalueCutoff = 0.2,
      readable = TRUE
    ) %>%
      as.data.frame() %>%
      mutate(Ontology = ..1, direction = ..2, comparison = ..3)
  ) %>%
  bind_rows() %>% 
  glimpse()


saveRDS(
  enrichGO_res,
  glue(
    "{wkdir}/data/input/temp/",
    "BAM_EnrichGO_res.rds"
  )
)


```

Visualizing the enriched GO terms for BAMs across groups.

```{r, eval=FALSE}

enrichGO_res <- readRDS(
  glue(
    "{wkdir}/data/input/temp/",
    "BAM_EnrichGO_res.rds"
  )
)
enrichGO_res %>% glimpse

for (direc in c("Upregulated", "Downregulated")) {
  for (comp in unique(enrichGO_res$comparison)) {
    message(glue("Plotting: {comp} - {direc}"))
    p <- enrichGO_res %>%
      filter(direction == direc) %>%
      filter(comparison == comp) %>%
      filter(qvalue <= 0.05) %>%
      slice_min(qvalue, n = 30) %>%
      slice_max(Count, n = 20, with_ties = FALSE) %>%
      mutate(Description = glue("{Description}:{Ontology}")) %>% 
      ggplot(aes(x = Count, y = fct_reorder(Description, Count))) +
      geom_col(width = 0.4, fill = "red") +
      theme_bw() +
      labs(
        x = "Gene Count", y = NULL,
        title = glue("{comp}: {direc}")
      )

    ggsave(
      glue(
        "{wkdir}/figures/temp/GO_analysis/",
        "{comp}_{direc}_{Sys.Date()}.png"
        ),
      p,
      width = 7, height = 5.5
    )
  }
}

```


::: panel-tabset
### Control WT - Control ASO 
![](../figures/temp/GO_analysis/Control%20WT%20-%20Control%20ASO_Upregulated_2024-06-03.png)
![](../figures/temp/GO_analysis/Control%20WT%20-%20Control%20ASO_Downregulated_2024-06-03.png)

### Control WT - Prebiotic WT
![](../figures/temp/GO_analysis/Control%20WT%20-%20Prebiotic%20WT_Upregulated_2024-06-03.png)
![](../figures/temp/GO_analysis/Control%20WT%20-%20Prebiotic%20WT_Downregulated_2024-06-03.png)

### Control WT - Prebiotic ASO
![](../figures/temp/GO_analysis/Control%20WT%20-%20Prebiotic%20ASO_Upregulated_2024-06-03.png)
![](../figures/temp/GO_analysis/Control%20WT%20-%20Prebiotic%20ASO_Downregulated_2024-06-03.png)

### Control ASO - Prebiotic WT
![](../figures/temp/GO_analysis/Control%20ASO%20-%20Prebiotic%20WT_Upregulated_2024-06-03.png)
![](../figures/temp/GO_analysis/Control%20ASO%20-%20Prebiotic%20WT_Downregulated_2024-06-03.png)

### Control ASO - Prebiotic ASO
![](../figures/temp/GO_analysis/Control%20ASO%20-%20Prebiotic%20ASO_Upregulated_2024-06-03.png)
![](../figures/temp/GO_analysis/Control%20ASO%20-%20Prebiotic%20ASO_Downregulated_2024-06-03.png)

### Prebiotic WT - Prebiotic ASO
![](../figures/temp/GO_analysis/Prebiotic%20WT%20-%20Prebiotic%20ASO_Upregulated_2024-06-03.png)
![](../figures/temp/GO_analysis/Prebiotic%20WT%20-%20Prebiotic%20ASO_Downregulated_2024-06-03.png)
:::



```{r}
sessionInfo()
```