---
title: "Loading Parsebio data into Seurat"
editor: source
author: "Joe Boktor"
date: '2025-04-10'
format: 
  html:
    self-contained: true
    code-fold: true
    code-tools: true
    fig-align: right
    fig-height: 8
    fig-width: 14 
---


```{r}
library(Seurat)
library(Matrix)
library(irlba)
library(scico)
library(tidyverse)
library(patchwork)
library(glue)
library(presto)
library(hdf5r)
library(SeuratDisk)
library(tictoc)
library(future)
library(strex)
library(cowplot)
library(aplot)
theme_set(theme_bw())

home_dir <- "/central/groups/MazmanianLab/joeB"
rnk_dir <- "/resnick/groups/mthomson/jboktor"
split_pipe_dir <- glue("{rnk_dir}/split-pipe_workflow")
wkdir <- glue("{home_dir}/snRNAseq_analysis/WILDRxSPF_brains")
anndata_path <- glue("{wkdir}/data/input/parse_bio_anndata")
data_path <- glue("{wkdir}/data/interim")
fig_path <- glue("{wkdir}/figures")
abca_color_pal <- readRDS(glue("{wkdir}/data/interim/abca_color_pal.rds"))

# Set memory limit for future
options(future.globals.maxSize = 200000 * 1024^2)  # Set to ~200GB
# Use more memory-efficient serialization
options(future.serialization.progressr = FALSE)
options(future.serialization.progressr.enable = FALSE)
future::plan(multisession, workers = 8)

```

```{r}
# Convenience functions
SaveFigure <- function(plots, name, type = "png", width, height, res){
  if(type == "png") {
    png(paste0(fig_path, name, ".", type),
      width = width, height = height, units = "in", res = 200)
  } else {
    pdf(paste0(fig_path, name, ".", type),
      width = width, height = height)
      }
      print(plots)
      dev.off()
}

SaveObject <- function(object, name, data_path){
  saveRDS(object, paste0(data_path, name, ".rds"))
}

ReadObject <- function(name, data_path){
  readRDS(paste0(data_path, name, ".rds"))
}
```


Loading Seurat object from h5ad file (this takes ~10 minutes for this 18GB file)

```{r}
tictoc::tic()
seurat_obj <- schard::h5ad2seurat(glue("{anndata_path}/wildrxspf_data.h5ad"))
seurat_obj
tictoc::toc()

```


```{r}
# # Setting our initial cell class to a single type, this will changer after clustering. 
# seurat_obj@meta.data$orig.ident <- factor(rep("seurat_obj", nrow(seurat_obj@meta.data)))
# Idents(seurat_obj) <- seurat_obj@meta.data$orig.ident

# SaveObject(seurat_obj, "seurat_obj_before_QC")
# seurat_obj <- ReadObject("seurat_obj_before_QC")
```


```{r}
# All nuclei have 0% mitochondrial reads
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^mt-")
seurat_obj[["percent.mt"]] %>% unique()
seurat_obj@meta.data %>% glimpse()

# plot <- VlnPlot(seurat_obj, #pt.size = 0.10,
#   features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# SaveFigure(plot, "vln_QC", width = 12, height = 6)

plot1 <- FeatureScatter(seurat_obj, 
                        feature1 = "nCount_RNA", 
                        feature2 = "percent.mt",
                        group.by = "sample")
plot2 <- FeatureScatter(seurat_obj, 
                       feature1 = "nCount_RNA", 
                       feature2 = "nFeature_RNA",
                       group.by = "sample") +
                geom_vline(xintercept = 1.7e5, linetype = "dashed", color = "red") +
                geom_hline(yintercept = 1.2e4, linetype = "dashed", color = "blue")

SaveFigure(plot1, "scatter_QC_with_MT", width = 9, height = 6)
SaveFigure(plot2, "scatter_QC", width = 9, height = 6)


```

Perform filtering based on the scatter plot. In this process we select for cells with fewer than 2% mitochondrial reads, less than 1.2e4 unique genes and less than 1.7e5 counts. Going from 414591 cells to 411875 cells.

```{r}
seurat_obj <- subset(seurat_obj, 
  subset = nFeature_RNA < 1.2e4 & nCount_RNA < 1.7e5 & percent.mt < 2)
seurat_obj
saveRDS(seurat_obj, 
    glue("{data_path}/seurat/seurat_obj_filtered_{Sys.Date()}.rds")
    )
```


Reading in filtered seurat object

```{r}
seurat_obj <- readRDS(glue(
  "{data_path}/seurat/seurat_obj_filtered_2025-04-22.rds")
  )
```


Normalizing the data

```{r}
future::plan("sequential")
tic()
seurat_obj <- NormalizeData(
    seurat_obj,
    normalization.method = "LogNormalize", scale.factor = 10000
)
toc()

```


```{r}
seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(seurat_obj), 10)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(seurat_obj)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
SaveFigure((plot1 + plot2), "var_features", width = 12, height = 6)
```

Performing PCA 


```{r}
tic()
# Scaling the data
seurat_obj <- ScaleData(seurat_obj)
# Running PCA
seurat_obj <- RunPCA(seurat_obj)
toc()

# Examine and visualize PCA results a few different ways
print(seurat_obj[["pca"]], dims = 1:5, nfeatures = 5)
```


```{r}
plot <- VizDimLoadings(seurat_obj, dims = 1:2, reduction = "pca")
SaveFigure(plot, "viz_PCA_loadings", width = 10, height = 8)
```

```{r}
plot <- DimPlot(seurat_obj, reduction = "pca", group.by = "orig.ident")
SaveFigure(plot, "pc1_2_scatter", width = 8, height = 6)
```

```{r}
# Image doesn't save as png unless fast = FALSE
plot <- DimHeatmap(seurat_obj, dims = 1, cells = 500, balanced = TRUE, fast = FALSE)
SaveFigure(plot, "dim_heatmap1", width = 8, height = 6)
```


```{r}
plot <- DimHeatmap(seurat_obj, dims = 1:15, cells = 500, balanced = TRUE, fast = FALSE)
SaveFigure(plot, "dim_heatmap1_15", width = 12, height = 18)
```


```{r}
plot <- ElbowPlot(seurat_obj,ndims = 50)
SaveFigure(plot, "PC_elbow_plot", width = 8, height = 10)
```


```{r}
seurat_obj <- JackStraw(seurat_obj, num.replicate = 100)
seurat_obj <- ScoreJackStraw(seurat_obj, dims = 1:20)
```


```{r}
seurat_obj <- FindNeighbors(seurat_obj, dims = 1:30)
seurat_obj <- FindClusters(seurat_obj, resolution = 0.30)
```


```{r}
seurat_obj <- BuildClusterTree(seurat_obj, reorder = TRUE, reorder.numeric = TRUE)

```


```{r}
future::plan(multisession, workers = 8)

seurat_obj <- RunUMAP(seurat_obj, dims = 1:30)
plot <- DimPlot(seurat_obj, reduction = "umap")
SaveFigure(plot, "umap_louvain_res_p3", width = 9, height = 8)
```


```{r}

# SaveObject(seurat_obj, "seurat_obj_clustered")
# seurat_obj <- ReadObject("seurat_obj_clustered")

# Idents(seurat_obj) %>% glimpse
```


```{r}
seurat_obj_markers <- FindAllMarkers(seurat_obj, min.pct = 0.25, logfc.threshold = 0.25)
seurat_obj_markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
```


```{r}
saveRDS(seurat_obj_markers, 
    glue("{data_path}/seurat/seurat_obj_markers_{Sys.Date()}.rds")
    )
saveRDS(seurat_obj, 
    glue("{data_path}/seurat/seurat_obj_clustered_{Sys.Date()}.rds")
    )

```

Reading in filtered & clustered seurat object

```{r}
seurat_obj <- readRDS(glue(
  "{data_path}/seurat/seurat_obj_clustered_2025-04-22.rds")
  )

seurat_obj@meta.data %>% glimpse()
```

Loading in MapMyCells cell type annotation

```{r}
cell_type_map <- read_csv(glue("{data_path}/MapMyCells_trial/output.csv"),
  skip = 4) %>% 
  glimpse()

cell_type_df <- cell_type_map %>% 
  column_to_rownames("cell_id") %>% 
  glimpse()

seurat_obj <- AddMetaData(seurat_obj, cell_type_df)
seurat_obj@meta.data %>% glimpse()

saveRDS(seurat_obj, 
    glue("{data_path}/seurat/seurat_obj_clustered_celltyped_{Sys.Date()}.rds")
    )

```

TOSS ME

```{r}
cell_type_map %>% glimpse()

cell_type_map$supertype_name %>% unique() %>% 
keep(grepl("OB", .))


cell_type_map$class_name %>% unique() %>% 

cell_type_map %>%
  filter(class_name == "03 OB-CR Glut") %>% 
  glimpse()
```



```{r}
seurat_obj@meta.data %>% glimpse()
# cell_type_map$class_bootstrapping_probability %>% summary()
# cell_type_map$class_bootstrapping_probability %>% hist()

cell_meta_annot_df <- seurat_obj@meta.data %>% 
  rownames_to_column("rowname") %>%
  left_join(cell_type_map, by = c("bc_wells" = "cell_id")) %>%
  column_to_rownames("rowname") %>%
  glimpse()

# cell_meta_annot_df %>% rownames() %>% head
# seurat_obj@meta.data <- cell_meta_annot_df
# seurat_obj@meta.data %>% rownames() %>% head

p_umap_celltype <- DimPlot(seurat_obj,
  group.by = "class_name",
  label = TRUE, 
  label.box = TRUE,
  raster = FALSE,
  repel = TRUE
) +
    coord_fixed() +
  scale_color_manual(values = abca_color_pal[["cluster_colors"]]) +
  scale_fill_manual(values = abca_color_pal[["cluster_colors"]])

ggsave(
  glue("{fig_path}/umaps/umap_mapmycells_{Sys.Date()}.png"),
  p_umap_celltype,
  width = 15, height = 10
)
```



Manual UMAP plotting

```{r}
seurat_obj <- readRDS(glue(
  "{data_path}/seurat/seurat_obj_clustered_celltyped_2025-05-12.rds")
  )

set.seed(123)
umap_coords <- Embeddings(seurat_obj, reduction = "umap") %>%
  cbind(seurat_obj@meta.data) %>% 
  mutate(tissue = str_after_last(sample, "_")) %>%
  mutate(group = str_before_first(sample, "_")) %>%
  slice_sample(prop = 1) %>%  # random shuffle row order
  glimpse()

p_celltype_umap <- umap_coords %>% 
  ggplot(aes(x = umap_1, y = umap_2, color = class_name)) +
  geom_point(size = 0.1, alpha = 0.2) +
  theme_void() +
  coord_fixed() +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  scale_color_manual(values = abca_color_pal[["cluster_colors"]])

ggsave(
  glue("{wkdir}/figures/umaps/umap_celltype_{Sys.Date()}.png"),
  p_celltype_umap,
  width = 9, height = 7
)

p_celltype_umap_group <- umap_coords %>% 
  # mutate(group = factor(group, levels = c("SPF" = "spf", "WildR" = "wildr"))) %>%
  ggplot(aes(x = umap_1, y = umap_2, color = group)) +
  geom_point(size = 0.1, alpha = 0.2) +
  theme_void() +
  coord_fixed() +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  ggsci::scale_color_d3()

ggsave(
  glue("{wkdir}/figures/umaps/umap_celltype_group_{Sys.Date()}.png"),
  p_celltype_umap_group,
  width = 9, height = 7
)



combined_umap <- p_celltype_umap + p_celltype_umap_group
# Save the combined plot
ggsave(
  glue("{wkdir}/figures/umaps/umap_celltype_combined_{Sys.Date()}.png"),
  combined_umap,
  width = 18, height = 7  # double the width for two plots side by side
)

# Just the legend
p_celltype_umap_4legend <- umap_coords %>% 
  slice_sample(prop = 1) %>%  # random shuffle row order
  ggplot(aes(x = umap_1, y = umap_2, color = class_name)) +
  geom_point(size = 0.1) +
  theme_void() +
  guides(color = guide_legend(override.aes = list(size = 4))) +
  scale_color_manual(values = abca_color_pal[["cluster_colors"]])

# Extract just the legend
legend_only <- get_legend(p_celltype_umap_4legend)
# Save the legend
ggsave(
  glue("{wkdir}/figures/umaps/umap_celltype_legend_{Sys.Date()}.png"),
  legend_only,
  width = 3, height = 8  # adjust these dimensions as needed for your legend
)


```


```{r}
meta_df <- seurat_obj@meta.data %>% glimpse()

celltype_counts_df <- meta_df %>% 
  mutate(group = str_before_first(sample, "_")) %>% 
  mutate(tissue = str_after_last(sample, "_")) %>% 
  group_by(class_name, sample, tissue) %>% 
  summarise(celltype_counts= n()) %>% 
  glimpse()

for (celltype in unique(celltype_counts_df$class_name)) {
  for (tis in c("AMY", "HYP")) {
    celltype_counts_df %>% 
      filter(class_name == celltype) %>% 
      filter(tissue == tis) %>% 
      glimpse()
  }
}


```



```{r}
# First, let's modify the data preparation to include group information
celltype_counts_df <- meta_df %>% 
  mutate(group = str_before_first(sample, "_")) %>% 
  mutate(tissue = str_after_last(sample, "_")) %>% 
  group_by(class_name, sample, tissue, group) %>% 
  summarise(celltype_counts= n()) %>% 
  group_by(sample) %>% 
  mutate(relab = celltype_counts/sum(celltype_counts) * 100) %>% 
  ungroup() %>% 
  glimpse()


View(celltype_counts_df)

# Now run the linear model for each cell type and tissue combination
results <- list()

for (celltype in unique(celltype_counts_df$class_name)) {
  for (tis in c("AMY", "HYP")) {
    # Filter data for current cell type and tissue
    current_data <- celltype_counts_df %>% 
      filter(class_name == celltype, tissue == tis)
    
    # Run linear model
    model_counts <- lm(celltype_counts ~ group, data = current_data)
    model_relab <- lm(relab ~ group, data = current_data)
    
    # Store results
    results[[paste(celltype, tis, sep = "_")]] <- summary(model)
  }
}

# Print results
for (name in names(results)) {
  cat("\nResults for:", name, "\n")
  print(results[[name]])
}



```



Testing for differences in relative abundance and total counts between groups

```{r}
seurat_obj <- readRDS(glue(
  "{data_path}/seurat/seurat_obj_clustered_celltyped_2025-05-12.rds")
  )

```



```{r}
# key sample metadata needs to be added into seurat_obj@meta.data
# Should move this up as early as possible
seurat_obj@meta.data$sample %>% sample(40)
seurat_obj@meta.data$region <- seurat_obj@meta.data$sample %>% str_after_last("_")
seurat_obj@meta.data$group <- seurat_obj@meta.data$sample %>% str_before_first("_")

seurat_obj@meta.data %>% glimpse()
count_df <- seurat_obj@meta.data %>% 
  as.data.frame() %>% 
  group_by(class_name, sample, region, group) %>% 
  dplyr::summarise(counts = n()) %>% 
  group_by(sample, region, group) %>% 
  dplyr::mutate(relab = counts/sum(counts) * 100) %>% 
  mutate(group_x_tissue = paste(group, region, sep = "_")) %>% 
  mutate(group_x_tissue = factor(group_x_tissue, 
    levels = c("SPF_AMY", "WildR_AMY", "SPF_HYP", "WildR_HYP"))) %>% 
  glimpse()

p_bar <- count_df %>% 
  ggplot(aes(x=sample, y=relab, fill = class_name)) +
  geom_bar(stat = "identity", color = "black", linewidth = 0.3) +
  scale_fill_manual(values = abca_color_pal[["cluster_colors"]]) +
  facet_wrap(~group_x_tissue, ncol = 2, scales = "free") +
  labs(x =NULL, y= "Relative Abundance (%)", fill = NULL) +
  theme_classic()  +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(
  glue("{fig_path}/celltype_plots/celltype_barplot_{Sys.Date()}.png"),
  p_bar,
  width = 9, height = 10
)

p_bar_counts <- count_df %>% 
  ggplot(aes(x=sample, y=counts, fill = class_name)) +
  geom_bar(stat = "identity", color = "black", linewidth = 0.3) +
  scale_fill_manual(values = abca_color_pal[["cluster_colors"]]) +
  facet_wrap(~group_x_tissue, ncol = 2, scales = "free_x") +
  labs(x =NULL, y= "Cell Counts", fill = NULL) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(
  glue("{fig_path}/celltype_plots/celltype_barplot_counts_{Sys.Date()}.png"),
  p_bar_counts,
  width = 9, height = 10
)
```

How does the cell type labeling compare with MapMyCells?

```{r}
seurat_obj@meta.data %>% glimpse()

tab <- table(
  Assigned = seurat_obj@meta.data$subclass_name,
  # Assigned = seurat_obj@meta.data$cluster_name,
  Clusters = seurat_obj@meta.data$seurat_clusters
)
p_cluster_annot_heatmap <-
  pheatmap::pheatmap(log10(tab + 1),
  # pheatmap::pheatmap(tab,
    color = colorRampPalette(c("white", "blue"))(10)
  )
ggsave(
  glue("{fig_path}/celltype_plots/celltype_heatmap_{Sys.Date()}.png"),
  p_cluster_annot_heatmap,
  # width = 9, height = 30
  width = 9, height = 7
)
```



```{r}
plot <- VlnPlot(seurat_obj, features = c("Itih3-GRCm39", "Cyp2j12-GRCm39"), group.by = "tree.ident")
SaveFigure(plot, "vln_exp1", width = 16, height = 8)
```


```{r}
# you can plot raw counts as well
plot <- VlnPlot(seurat_obj, features = c("Itih3-GRCm39", "Cyp2j12-GRCm39"), slot = "counts", log = TRUE)
SaveFigure(plot, "vln_exp2", width = 16, height = 8)
```


```{r}
top5 <- seurat_obj_markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
to_plot <- unique(top5$gene)
plot <- DotPlot(seurat_obj, features = to_plot, group.by = "tree.ident") + coord_flip()
SaveFigure(plot, "dplot_top5", width = 16, height = 25)
```


```{r}
# Plotting seurat_obj markers
markers <- c("Il7r", "CCR7", "S100A4", "CD14", "LYZ", "MS4A1", "CD8A",
"FCGR3A", "MS4A7", "GNLY", "NKG7", "FCER1A", "CST3", "PPBP", "IGHA2", "ZNF385D")

seurat_obj_markers <- FeaturePlot(seurat_obj, features = markers)
SaveFigure(seurat_obj_markers, "seurat_obj_markers", width = 30, height = 30)
```


```{r}
plot <- DimPlot(seurat_obj, reduction = "umap", label = TRUE) + NoAxes() + NoLegend
SaveFigure(plot, "umap_louvain_nolegend", width = 8, height = 7)
```


```{r}
# New IDs from official tutorial
new_ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T",
  "FCGR3A+ Mono", "NK", "DC", "IGHA", "ZNF385D")

new_id_list <- list(Naive.CD4.T = c(17,19,20), Memory.CD4.T = c(15,16,18),
  CD14.Mono = c(2), B = c(5:7), CD8.T = 13, FCGR3A.Mono = 1, NK = c(10:12,14),
  DC = c(3,8), IGHA = 4, ZNF385D = 9)

for (i in 1:length(new_id_list)) {
  ind <- which(seurat_obj@meta.data$tree.ident %in% new_id_list[[i]])
  seurat_obj@meta.data$collapsed[ind] <- names(new_id_list)[i]
}

seurat_obj@meta.data$collapsed <- factor(
  seurat_obj@meta.data$collapsed, levels = names(new_id_list), ordered = TRUE)
Idents(seurat_obj) <- seurat_obj@meta.data$collapsed

names(new_ids) <- levels(seurat_obj)
seurat_obj <- RenameIdents(seurat_obj, new_ids)

plot <- DimPlot(seurat_obj, reduction = "umap", label = TRUE) + NoAxes() + NoLegend()
SaveFigure(plot, "umap_louvain_nolegend_names", width = 8, height = 7)
```
